# This code performs CTS analysis on the expressional data of scRNAseq, using BioTIP (v.1.1.0).
# THe GEO-downloaded FPKM values were normalized by monocle.
# Generated by:  Xinan H Yang (xyang2_at_uchicago.edu)
# Last update: 09/29/2020


library(BioTIP)
packageVersion("BioTIP")  #[1] '1.3.0'


setwd("F:/projects/BioTIP/doc/2020_/Applications/")
#source("../../../source/BioTIP_update_3.3_02282020.R")
#source("F:/projects/BioTIP/R_function/BioTIP-master/R/BioTIP_update_4_09282020.R")
library(stringr)
library(psych)
library(igraph)

###############################################################
#### read the transcriptome data matrix and sample information
###############################################################  
  
load(file="F:/projects/BioTIP/doc/2020_/Applications/input.Rdata/GSE52583/GSE52583_monocle_counts.RData")
dim(dat) #[1] 10359   196

load(file="F:/projects/BioTIP/doc/2020_/Applications/input.Rdata/GSE52583/GSE52583_cli.RData")
dim(cli) # 196  21
 
# check if sample IDs in the 'cli' matris are identical to the sample names in the 'dat' matrix  
all(rownames(cli) == colnames(dat))  # TRUE

#####################################
#### test 1) BioTIP identification with age-based cell grouping
#######################################
 samplesL <- split(rownames(cli),f = cli$age)
 tmp <- lapply(samplesL, length)
 if(any(tmp<4))  samplesL <- samplesL[-which(tmp<4)] 
 
 
 table(cli$cell_type, cli$State)
 #                1  2  3  4  5
 # 14.5_NA       44  0  0  1  0
 # 16.5_NA       25  0  0  2  0
 # 18.5_BP        0  5  3  5  0
 # 18.5_ciliated  3  0  0  0  0
 # 18.5_Clara     2  3  0  6  0
 # 18.5_AT1       2 24  4  5  6
 # 18.5_AT2       0  0  0 12  0
 # Adult_AT2      1  0  0 43  0
 table(cli$cell_type, cli$Cluster)
 #                1  2  3
 # 14.5_NA        1 44  0
 # 16.5_NA        0 27  0
 # 18.5_BP        5  0  8
 # 18.5_ciliated  0  3  0
 # 18.5_Clara     4  0  7
 # 18.5_AT1       4  0 37
 # 18.5_AT2      12  0  0
 # Adult_AT2     43  0  1

########################################################################################################### 
####### 1) whether random genes using old IC() or new BioTIO() have bias for tipping-point identification 
########################################################################################################### 
# DO NOT REPEAT , takes a while
 #  B=1000
#  output='Ic'
#  source("../../../source/R_script_4_ICsimulation.R")
#  
#  output='PCCg'
#  source("../../../source/R_script_4_ICsimulation.R")
# 
#  output='PCCs'
#  source("../../../source/R_script_4_ICsimulation.R")
#  getwd()
# # "F:/projects/BioTIP/result/GSE52583/log2_all"
#  
#   
#  
 ####### 1) whether random genes using old IC() or new BioTIO() have bias for tipping-point identification 
 B <-  1000  # 1k simulation times
 outputs <- c('Ic','PCCg','PCCs')
 
 n <- 1000 # 1k random genes
 n.smallest.sample <- min(unlist(lapply(samplesL, length)))
 
 if(any(is.na(dat))) use="pairwise.complete.obs" else use="everything"

 for(j in outputs) 
 {
    output <- j
    
 set.seed(2020)
 for(k in c("cor","BioTIP"))
 {
   method <- ifelse(k=='cor', "IC","BioTIP")

   ## how about permutate genes?
   IC_simu <- array(dim=c(length(samplesL), B))
   rownames(IC_simu) <- names(samplesL)
   IC_simu <- simulation_Ic(n, samplesL, counts=dat, B=B, fun=k, output=output, use=use)
   save(IC_simu, file=paste0(output,"_Simu",method,"_fr_",nrow(dat),"G.RData"), compress=T)
   
   
   
   ########  how about and then permutate genes, but also randomly select n.smallest.sample cells per age ?
  # set.seed(2020)
   IC_simu <- array(dim=c(length(samplesL), B))
   rownames(IC_simu) <- names(samplesL)
   samples <- lapply(samplesL, function(x) sample(x, n.smallest.sample))
   IC_simu <- simulation_Ic(n, samples, dat, B=B, fun=k, output=output, use=use)
   save(IC_simu, n, file=paste0(output,"_Simu",method,"_fr_",nrow(dat),"G_",n.smallest.sample,"cellPerAge.RData"), compress=T)
 } 
 }   
   
   # ######## sample-label shuffling also gene permutation #######################
 # for(k in c("cor","BioTIP"))
 # {
 #    method <- ifelse(k=='cor', "IC","BioTIP")
 #  
   # IC_simu <- array(dim=c(length(samplesL), B))
   # rownames(IC_simu) <- names(samplesL)
   # simu <- matrix(nrow=length(samplesL), ncol=B)
   # rownames(simu) = names(samplesL)
   # for(j in 1:length(samplesL)) {
   #     cat(".")
   #     g <- sample(rownames(dat), n )
   #     IC_simu[j,] <- simulation_Ic_sample(dat, sampleNo=length(samplesL[[j]]), Ic= NULL,
   #                                      genes=g,  B=B,
   #                                      fun=k, plot=FALSE, output=output, use=use)
   #    cat("\n")
   #    }
   # save(IC_simu, file=paste0(output,"_Simu",method,"_shulfLabelling_fr_",nrow(dat),"G.RData"), compress=T)
   
   
   # ###########  sample-label shuffling also gene permutation, but n.smallest.sample cells per age  random select samples beyond staage  #######
   # 
   # IC_simu <- array(dim=c(length(samplesL), B))
   # rownames(IC_simu) <- names(samplesL)
   # simu <- matrix(nrow=length(samplesL), ncol=B)# 4 500
   # rownames(simu) = names(samplesL)
   # for(j in 1:length(samplesL)) {
   #     cat(".")
   #     g <- sample(rownames(dat), n[i] )
   #     IC_simu <- simulation_Ic_sample(dat, sampleNo=n.smallest.sample, Ic= NULL,
   #                                      genes=g,  B=B,
   #                                      fun=k, plot=FALSE, output=output, use=use)
   #     cat("\n")
   #     }
   # dimnames(IC_simu)[[3]] <- n
   # save(IC_simu, file=paste0(output,"_Simu",method,"_shulfLabelling_fr_",nrow(dat),"G_",n.smallest.sample,"cellPerAge.RData"), compress=T)
   
# }

 
 
#########################################################################
### 1.3) whether we can estimate the existence of a tipping point??
#########################################################################
 dim(dat) #dim(dat) # [1] 10359   196
 

library(MASS)
library(mixtools)
# https://cran.r-project.org/web/packages/mixtools/vignettes/mixtools.pdf

#   datafold <- "F:/projects/BioTIP/result/GSE52583/log2_all/"
   datafold <- "F:/projects/BioTIP/doc/2020_/Applications/"
   load(file=paste0(datafold,"Ic_SimuIc_fr_10359G.RData")) 
   dim(IC_simu)   #   4 1000    
   Ic_random = IC_simu#[,,3]
   
   load(file=paste0(datafold,"Ic_SimuBioTIP_fr_10359G.RData")) 
   dim(IC_simu)   #   4 1000    
   Ic.new_random = IC_simu#[,,3]
   
   load(file=paste0(datafold,"Ic_SimuIc_fr_10359G_27cellPerAge.RData"))  # _shulfLabelling
   dim(IC_simu)   #   4 1000    
   Ic_random.27cell = IC_simu#[,,3]
   
   load(file=paste0(datafold,"Ic_SimuBioTIP_fr_10359G_27cellPerAge.RData")) 
   dim(IC_simu)   #   4 1000    
   Ic.new_random.27cell = IC_simu#[,,3]
   
par(mfrow=c(2,2))
plot_Ic_Simulation(apply(Ic_random.27cell,1,median), Ic_random.27cell, las = 2, ylab="Ic",
                   main="27cell, 1000 g-permutations",
                   fun="boxplot", which2point= 2)
plot_Ic_Simulation(apply(Ic.new_random.27cell,1,median), Ic.new_random.27cell, las = 2, ylab="Ic*",
                   main="27cell,  1000g-permutations",
                   fun="boxplot", which2point= 2)

plot_Ic_Simulation(apply(Ic_random,1,median), Ic_random, las = 2, ylab="Ic",
                   main="1000 cell-permutations",
                   fun="boxplot", which2point= 2)
plot_Ic_Simulation(apply(Ic.new_random,1,median), Ic.new_random, las = 2, ylab="Ic*",
                   main="1000 gene-permutations",
                   fun="boxplot", which2point= 2)
dev.copy2pdf(file='ROutputFigs/GSE52583/Icvs_Ic.new_1krandom.gene.pdf')

###########################################################
## Compare PCCs and PCCg into details 
datafold2 <- "F:/projects/BioTIP/result/GSE52583/log2_all/"
datafold <- "F:/projects/BioTIP/doc/2020_/Applications/"

load(file=paste0(datafold,"PCCg_SimuIc_fr_10359G.RData")) 
dim(IC_simu)   #   4 1000    # was 3 the results of random 20, 200 and 1000 genes
PCCg_random = IC_simu#[,,3]
load(file=paste0(datafold2,"PCCg_SimuBioTIP_fr_10359G.RData")) 
dim(IC_simu)   #   4 1000    
PCCg.new_random = IC_simu[,,3]

load(file=paste0(datafold,"PCCs_SimuIc_fr_10359G.RData")) 
dim(IC_simu)   #   4 1000    
PCCs_random = IC_simu#[,,3]
load(file=paste0(datafold2,"PCCs_SimuBioTIP_fr_10359G.RData")) 
dim(IC_simu)   #   4 1000   
PCCs.new_random = IC_simu[,,3]

load(file=paste0(datafold,"PCCg_SimuIc_fr_10359G_27cellPerAge.RData"))  # _shulfLabelling
dim(IC_simu)   #   4 1000    
PCCg_random.27cell = IC_simu#[,,3]
load(file=paste0(datafold2,"PCCg_SimuBioTIP_fr_10359G_27cellPerAge.RData")) 
dim(IC_simu)   #   4 1000   
PCCg.new_random.27cell = IC_simu[,,3]

load(file=paste0(datafold,"PCCs_SimuIc_fr_10359G_27cellPerAge.RData"))  # _shulfLabelling
dim(IC_simu)   #   4 1000   
PCCs_random.27cell = IC_simu#[,,3]
load(file=paste0(datafold2,"PCCs_SimuBioTIP_fr_10359G_27cellPerAge.RData")) 
dim(IC_simu)   #   4 1000   
PCCs.new_random.27cell = IC_simu[,,3]

par(mfrow=c(2,4))
plot_Ic_Simulation(apply(PCCg_random.27cell,1,median), PCCg_random.27cell, las = 2, ylab="PCCg",
                   main="27cell, 1000 g-permutations",
                   fun="boxplot", which2point= 2)
plot_Ic_Simulation(apply(PCCs_random.27cell,1,median), PCCs_random.27cell, las = 2, ylab="PCCs",
                   main="27cell, 1000 g-permutations",
                   fun="boxplot", which2point= 2)
plot_Ic_Simulation(apply(PCCg.new_random.27cell,1,median), PCCg.new_random.27cell, las = 2, ylab="PCCg*",
                   main="27cell,  1000g-permutations",
                   fun="boxplot", which2point= 2)
plot_Ic_Simulation(apply(PCCs.new_random.27cell,1,median), PCCs.new_random.27cell, las = 2, ylab="PCCs*",
                   main="27cell,  1000g-permutations",
                   fun="boxplot", which2point= 2)

plot_Ic_Simulation(apply(PCCg_random,1,median), PCCg_random, las = 2, ylab="PCCg",
                   main="1000 cell-permutations",
                   fun="boxplot", which2point= 2)
plot_Ic_Simulation(apply(PCCs_random,1,median), PCCs_random, las = 2, ylab="PCCs",
                   main="1000 cell-permutations",
                   fun="boxplot", which2point= 2)
plot_Ic_Simulation(apply(PCCg.new_random,1,median), PCCg.new_random, las = 2, ylab="PCCg*",
                   main="1000 gene-permutations",
                   fun="boxplot", which2point= 2)
plot_Ic_Simulation(apply(PCCs.new_random,1,median), PCCs.new_random, las = 2, ylab="PCCs*",
                   main="1000 gene-permutations",
                   fun="boxplot", which2point= 2)

dev.copy2pdf(file='ROutputFigs/GSE52583/PCCvs_PCC.new_1krandom.gene.pdf')





#########################################################
####### 2) find CTS ######################
# Pre-selection Transcript

cut.preselect = 0.05
cut.fdr = 0.1 # 0.05
cut.minsize = 30

test <- sd_selection(dat, samplesL, cut.preselect)
names(test)
head(test[["16.5"]])

# Network Partition
 
igraphL <- getNetwork(test, fdr = cut.fdr)
# 14.5:517 nodes
# 16.5:517 nodes
# 18.5:521 nodes
# Adult:517 nodes


cluster <- getCluster_methods(igraphL)
names(cluster)
# [1] "14.5"  "16.5"  "18.5"  "Adult"

head(cluster[[1]])


# 2.1) Identifying Dynamic Network Biomodule using old MCI-score ########
###########################################################################
membersL_noweight <- getMCI(cluster,test, adjust.size = FALSE, fun = "cor")
names(membersL_noweight)
#[1] "members" "MCI"     "sd"      "PCC"     "PCCo"  
# plotBar_MCI(membersL_noweight, ylim = c(0,150), minsize = cut.minsize)
# dev.copy2pdf(file=paste0("MCI_bar_preselect", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"))
#the numbers in the parenthesis: they are total number of clusters, no control of the cell (sample) size per  cluster

# get the statistics
maxMCIms <- getMaxMCImember(membersL_noweight[["members"]],membersL_noweight[["MCI"]],min =cut.minsize)
names(maxMCIms)
#[1] "idx"     "members"
head(maxMCIms[['idx']])  #  
# 14.5  16.5  18.5 Adult 
# 24     6     5     1 
head(maxMCIms[['members']][['18.5']])
#[1] "0610009B14Rik" "AW495222"      "Abp1"          "Acaa1a"       
#[5] "Acaca"         "Acap3"   

biomodules = getMaxStats(membersL_noweight[['members']], maxMCIms[[1]])
maxMCI = getMaxStats(membersL_noweight[['MCI']],maxMCIms[[1]])
head(maxMCI)
#  14.5      16.5      18.5     Adult 
#22.72876  29.81868 131.83505  14.68497 
CTS.E18 = getCTS(maxMCI, maxMCIms[[2]])
# Length: 205

maxMCI.2nd = getMaxStats(membersL_noweight[['MCI']][-3],maxMCIms[["idx"]][-3])
head(maxMCI.2nd)
#    14.5     16.5    Adult 
# 22.72876 29.81868 14.68497
CTS.E16 = getCTS(maxMCI.2nd, maxMCIms[["members"]][-3])
#Length: 69



# save for repeating and plot (also to Andrew)
save(CTS.E16, CTS.E18, file="outPut.Rdata/GSE52583/Treutlein2014_scRNAseq_lung_CTS_0.05_fdr0.1_minsize30.RData") 



######################################
## estimate significance
########################################

C = 1000
load(file=paste0("GSE52583_GenePermutation_",C,"CTS", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))
simuMCI.E18 <- simulationMCI(length(CTS.E18), samplesL, dat,  B=C, fun='cor')
simuMCI.E18_2nd <- simulationMCI(length(CTS.E18_2nd), samplesL, dat,  B=C)
simuMCI.E16 <- simulationMCI(length(CTS.E16), samplesL, dat,  B=C, fun='cor')
simuMCI.E14<- simulationMCI(length(CTS.E14), samplesL, dat,  B=C, fun='cor')

#save(simuMCI.E14,simuMCI.E16, simuMCI.E18, simuMCI.E18_2nd, file=paste0("GSE52583_GenePermutation_",C,"CTS", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))


#### plot ###################
# load(file=paste0("Shrinked",PCC_gene.target,".Treutlein2014_scRNAseq_lung_CTS_", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))
# load(file=paste0("GSE52583_GenePermutation_",C,"CTS", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))
par(mfrow=c(1,4))
observation = membersL_noweight[["MCI"]][['14.5']]['24']; names(observation) <- '14.5'
plot_MCI_Simulation(observation, simuMCI.E14, ylim=c(0,100),which2point='14.5',las=2,
                    main=paste(length(CTS.E14), "E14-selected genes", "\n","vs. ",C, "times of gene-permutation"))
observation =  membersL_noweight[["MCI"]][['16.5']]['6']; names(observation) <- '16.5'
plot_MCI_Simulation(observation, simuMCI.E16, ylim=c(0,60),which2point='16.5',las=2,
                    main=paste(length(CTS.E16), "E16-selected genes", "\n","vs. ",C, "times of gene-permutation"))
observation =  membersL_noweight[["MCI"]][['18.5']]['5']; names(observation) <- '18.5'      # 441.80687
plot_MCI_Simulation(observation, simuMCI.E18,  ylim=c(0,150), which2point='18.5',las=2,
                    main=paste(length(CTS.E18), "E18-selected genes", "\n","vs. ",C, "times of gene-permutation"))
observation = membersL_noweight[["MCI"]][['18.5']]['1']; names(observation) <- '18.5'
plot_MCI_Simulation(observation, simuMCI.E18_2nd,  ylim=c(0,20), which2point='18.5',las=2,
                    main=paste(length(CTS.E18_2nd), "E18-selected genes", "\n","vs. ",C, "times of gene-permutation"))

dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/MCI_GenePermutation_",C,"CTS", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"))




########################################################################################
####### 3)  Finding Tipping Point and verify using new IC score  ######################
########################################################################################


which2points <- c('16.5','18.5')
CTS <- list('16.5'=CTS.E16, '18.5'=CTS.E18)

set.seed(2020)
for(i in 1:length(CTS)) 
{
  n <- length(CTS[[i]])
  BioTIP_E18 <- getIc(dat, samplesL, genes=CTS[[i]], fun="BioTIP", PCC_sample.target='average')
  simuBioTIP_g <- simulation_Ic(n, samplesL, dat, B=C,
                                fun="BioTIP")
  simuBioTIP_s <- matrix(nrow=length(samplesL), ncol=C)
  rownames(simuBioTIP_s) = names(samplesL)
  for(j in 1:length(samplesL)) {
    n <- length(samplesL[[j]])  # of cells
    simuBioTIP_s[j,] <- simulation_Ic_sample(dat, sampleNo=n, Ic=BioTIP_E18[which2points[i]],
                                             genes=CTS[[i]],  B=C,
                                             fun="BioTIP")
  }
  
  
  par(mfrow=c(2,2))
  plot_Ic_Simulation(BioTIP_E18, simuBioTIP_g,las = 2, ylab="Ic*", ylim=c(0,4),
                     main=paste(length(CTS[[i]]),names(CTS[i]), "genes", "\n","vs. ",C, "times of gene-permutation"),
                     fun="boxplot")
  plot_SS_Simulation(BioTIP_E18, simuBioTIP_g, xlim=c(0,4),
                     main = paste("Larget - 2nd largest",length(CTS[[i]]),"genes"), #which2point="18.5", 
                     ylab="Ic*")
  
  plot_Ic_Simulation(BioTIP_E18, simuBioTIP_s,las = 2, ylab="Ic*", ylim=c(0,30),
                     main=paste(length(CTS[[i]]),names(CTS[i]), "genes", "\n","vs. ",C, "times of cell-permutation"),
                     fun="boxplot")
  plot_SS_Simulation(BioTIP_E18, simuBioTIP_s, xlim=c(0,100),
                     main = paste("Larget - 2nd largest",length(CTS[[i]]),"genes"), #which2point="18.5", 
                     ylab="Ic*")
  
  dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/BioTIP_vsSimulation_",length(CTS[[i]]),"CTS.pdf"))
}




### how about  without shrinkage, the old Ic-score  
for(i in 1:length(CTS)) 
{
  n <- length(CTS[[i]])
  BioTIP_E18 <- getIc(dat, samplesL, genes=CTS[[i]], fun="cor", PCC_sample.target='average')
  simuBioTIP_g <- simulation_Ic(n, samplesL,dat, B=C,
                                fun="cor")
  simuBioTIP_s <- matrix(nrow=length(samplesL), ncol=C)
  rownames(simuBioTIP_s) = names(samplesL)
  for(j in 1:length(samplesL)) {
    n <- length(samplesL[[j]])  # of cells
    simuBioTIP_s[j,] <- simulation_Ic_sample(dat, sampleNo=n, Ic=BioTIP_E18[which2points[i]],
                                             genes=CTS[[i]],  B=C,
                                             fun="cor")
  }
  
  
  par(mfrow=c(2,2))
  plot_Ic_Simulation(BioTIP_E18, simuBioTIP_g,las = 2, ylab="Ic",
                     main=paste(length(CTS[[i]]),names(CTS[i]), "genes", "\n","vs. ",C, "times of gene-permutation"),
                     fun="boxplot", which2point=which2points[i])
  plot_SS_Simulation(BioTIP_E18, simuBioTIP_g, 
                     main = paste("Larget - 2nd largest",length(CTS[[i]]),"genes"), #which2point="18.5", 
                     ylab="Ic")
  
  plot_Ic_Simulation(BioTIP_E18, simuBioTIP_s,las = 2, ylab="Ic",
                     main=paste(length(CTS[[i]]),names(CTS[i]), "genes", "\n","vs. ",C, "times of cell-permutation"),
                     fun="boxplot", which2point=which2points[i])
  plot_SS_Simulation(BioTIP_E18, simuBioTIP_s, 
                     main = paste("Larget - 2nd largest",length(CTS[[i]]),"genes"), #which2point="18.5", 
                     ylab="Ic")
  
  dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/Ic_vsSimulation_",length(CTS[[i]]),"CTS.pdf"))
}







