# This code performs CTS analysis on the expressional data of scRNAseq, using BioTIP (v.1.1.1).
# THe GEO-downloaded FPKM values were normalized by monocle.
# Generated by:  Xinan H Yang (xyang2_at_uchicago.edu)
# Last update: 03/09/2020



library(igraph)
library(BioTIP)
packageVersion("BioTIP")  #[1] '1.1.1'


setwd("F:/projects/BioTIP/doc/2020_/Applications/")
#source("../../../source/BioTIP_update_3.3_02282020.R")

###############################################################
#### read the transcriptome data matrix and sample information
###############################################################  
  
load(file="F:/projects/BioTIP/doc/2020_/Applications/input.Rdata/GSE52583/GSE52583_monocle_counts.RData")
dim(dat)

load(file="F:/projects/BioTIP/doc/2020_/Applications/input.Rdata/GSE52583/GSE52583_cli.RData")
dim(cli)
 
# check if sample IDs in the 'cli' matris are identical to the sample names in the 'dat' matrix  
all(rownames(cli) == colnames(dat))  # TRUE

#####################################
#### test 1) BioTIP identification with age-based cell grouping
#######################################
 samplesL <- split(rownames(cli),f = cli$age)
 tmp <- lapply(samplesL, length)
 if(any(tmp<4))  samplesL <- samplesL[-which(tmp<4)] 
 
 
 table(cli$cell_type, cli$State)
 #                1  2  3  4  5
 # 14.5_NA       44  0  0  1  0
 # 16.5_NA       25  0  0  2  0
 # 18.5_BP        0  5  3  5  0
 # 18.5_ciliated  3  0  0  0  0
 # 18.5_Clara     2  3  0  6  0
 # 18.5_AT1       2 24  4  5  6
 # 18.5_AT2       0  0  0 12  0
 # Adult_AT2      1  0  0 43  0
 table(cli$cell_type, cli$Cluster)
 #                1  2  3
 # 14.5_NA        1 44  0
 # 16.5_NA        0 27  0
 # 18.5_BP        5  0  8
 # 18.5_ciliated  0  3  0
 # 18.5_Clara     4  0  7
 # 18.5_AT1       4  0 37
 # 18.5_AT2      12  0  0
 # Adult_AT2     43  0  1

########################################################################################################### 
####### 1) whether random genes using old IC() or new BioTIO() have bias for tipping-point identification 
########################################################################################################### 
 B=1000
 output='Ic'
 source("../../source/R_script_4_ICsimulation.R")
 
 output='PCCg'
 source("../../source/R_script_4_ICsimulation.R")

 output='PCCs'
 source("../../source/R_script_4_ICsimulation.R")
 
 hist(log2(dat[dat>=0.001]),100)
 dat[dat<0.001] <- NA
 hist(log2(dat),100) 
 dat <- log2(dat)
 
 setwd('F:/projects/BioTIP/result/GSE52583/log2_all')
 hist(log2(dat),100) 
 dev.copy2pdf(file="hist_log2.larger.than.0.001.pdf")
 

#########################################################################
### 1.3) whether we can estimate the existence of a tipping point??
#########################################################################
n <- c(20, 200, 500, 1000, 2000)
 dat <- exprs(annot.cds)
 dim(dat) #dim(dat) # [1] 10359   196
 

library(MASS)
library(mixtools)
# https://cran.r-project.org/web/packages/mixtools/vignettes/mixtools.pdf

#https://www.r-bloggers.com/palettes-in-r/

 colors <-c(#"black",
            "#A7A7A7",
           "dodgerblue",
         #  "firebrick",
           "gold",
           "forestgreen"
           )

## select the cut of Bipartite distribution 
 par(mfrow=c(1,3))
 for(i in c(20,200,500))
 {
   load(file=paste0("PCCsample2one/simuBioTIP_",i,"g_fr_10359G.RData")) 
   dim(simuBioTIP)   #   4 500
   tmp <- matrix(nrow=4, ncol=ncol(simuBioTIP))
   rownames(tmp) <- c('BioTIP_1st','BioTIP_2nd','state_1st','state_2nd')
   
   tmp[1:2,] <- apply(simuBioTIP, MARGIN = 2, function(x) sort(x, decreasing = T)[2:1])
   tmp[3:4,] <- apply(simuBioTIP, MARGIN = 2, function(x) order(x, decreasing = T)[2:1]) 
   diff_SS <- apply(tmp[1:2,], MARGIN = 2, diff)
   hist(diff_SS, 100, main=paste(i, " random genes,","\n", "S = 1st - 2nd Largest BioTIP Scores", sep = ""))
   abline(v = quantile(diff_SS,probs = seq(0,1,0.05))['95%'], col = 'red', lty=2, lwd = 2)
   
   }
 
 n  # 20, 200, 1000, 2000
## only show 20, 200, 2000 for simplicity
par(mfrow=c(1,3))

for(i in c(20,200,1000))
{
#  load(file=paste0("PCCsample2one/simuBioTIP_",i,"g_fr_10359G.RData")) 
  load(file=paste0("simuBioTIP_",i,"g_fr_10359G.RData")) 
  dim(simuBioTIP)   #   4 500
  tmp <- matrix(nrow=4, ncol=ncol(simuBioTIP))
  rownames(tmp) <- c('BioTIP_1st','BioTIP_2nd','state_1st','state_2nd')
  
  # png(paste("../Output/February 3 Presentation/Plots/scRNA seq ", i, " Random Labels (1) - (2) Distribution.png", sep = ""), width = 1200, height = 800)
  #  # dev.off()
  # rank the results per run
  # eval(parse(text = paste("diff_SS <- apply(apply(IC_rand_SS_", i, ", MARGIN = 2, function(x) sort(x, decreasing = T)[2:1]), MARGIN = 2, diff)", sep = "")))
  tmp[1:2,] <- apply(simuBioTIP, MARGIN = 2, function(x) sort(x, decreasing = T)[2:1])
  tmp[3:4,] <- apply(simuBioTIP, MARGIN = 2, function(x) order(x, decreasing = T)[2:1]) 
  diff_SS <- apply(tmp[1:2,], MARGIN = 2, diff)
 
  #  fit <- fitdistr(diff_SS, densfun="normal")  #Distributions "beta", "cauchy", "chi-squared", "exponential", "gamma", "geometric", "log-normal", "lognormal", "logistic", "negative binomial", "normal", "Poisson", "t" and "weibull" are recognised,
  # #  fit
  # #  shape         rate   
  # #  0.75337770   8.09624263 
  # #  (0.04088053) (0.60582297)
  # (fit.coef <- coef(fit))
  # #  mean         sd 
  # #  0.09307111 0.13998582 
  # cut = quantile(diff_SS,probs = seq(0,1,0.25))['75%']
  
  # the values are from two regimes - low and high - and assuming that the underlying process is normal, I used the mixtools package to fit a bimodal distribution:
  fit2 <- normalmixEM(diff_SS,   k=2)
#   plot(fit2, density=TRUE, cex.axis=1.4, cex.lab=1.4, cex.main=1.8,
#        main2=paste(i, " random genes"), xlab2=paste("S = 1st - 2nd Largest BioTIP Scores", sep = ""))
# fit[c("lambda", "mu", "sigma")]
# summary(fit2)
 
  # sigma: The final standard deviations 
#  cut = mean(fit2$mu)  # mu: The final mean parameters
   cut <- max(fit2$mu)
  
  diff_SS_grouped <- split(diff_SS, f=tmp['state_1st',])
  names(diff_SS_grouped) <- rownames(simuBioTIP)
  diff_SS_grouped$all <- diff_SS

  # Kernel Density Estimation 
  density_grouped <- lapply(diff_SS_grouped,density)
 
    
  ymax <- lapply(density_grouped, function(x) max(x$y)) 
    
  plot(density_grouped$all, type = 'l', lwd = 2, col = 'black', 
        main = paste(i, " random genes,","\n", "S = 1st - 2nd Largest BioTIP Scores", sep = ""), 
        xlim = c(-.05, 1.1) , 
        ylim = c(0, .1 + max(c(density_diff_SS$y,unlist(ymax) ))), 
        cex.main = 1, cex.lab = 1.5)
  for(j in 1:(length(density_grouped)-1)) points(density_grouped[[j]], type = 'l', lwd = 2, col = colors[j])
   
  abline(v = cut, col = 'red', lty=2, lwd = 2)
  order.to.show <- c(nrow(simuBioTIP)+1, 1:nrow(simuBioTIP))
  legend("topright", legend =  paste(c("All",rownames(simuBioTIP)), "higher S: ",
                  unlist(lapply(diff_SS_grouped, function(x) round(length(which(x>=cut))/length(x),2)*100))[order.to.show], "%"),
              fill = c( 'black',colors), bty = 'n', cex = 1) 
  
}

#dev.copy2pdf(file="PCCsample2one/Predict_tippingpoint.pdf") 
dev.copy2pdf(file="Predict_tippingpoint.pdf") 



####### 2) find biomarker ######################
# Pre-selection Transcript

cut.preselect = 0.05
cut.fdr = 0.1 # 0.05
cut.minsize = 30

test <- sd_selection(dat, samplesL, cut.preselect)
names(test)
head(test[["16.5"]])

# Network Partition
 
igraphL <- getNetwork(test, fdr = cut.fdr)
# 14.5:517 nodes
# 16.5:517 nodes
# 18.5:521 nodes
# Adult:517 nodes


cluster <- getCluster_methods(igraphL)
names(cluster)
# [1] "14.5"  "16.5"  "18.5"  "Adult"

head(cluster[[1]])


# 2.1) Identifying Dynamic Network Biomodule using old MCI-score ########
###########################################################################
membersL_noweight <- getMCI(cluster,test, adjust.size = FALSE, fun = "cor")
names(membersL_noweight)
#[1] "members" "MCI"     "sd"      "PCC"     "PCCo"  
# plotBar_MCI(membersL_noweight, ylim = c(0,150), minsize = cut.minsize)
# dev.copy2pdf(file=paste0("MCI_bar_preselect", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"))
#the numbers in the parenthesis: they are total number of clusters, no control of the cell (sample) size per  cluster

# get the statistics
maxMCIms <- getMaxMCImember(membersL_noweight[["members"]],membersL_noweight[["MCI"]],min =cut.minsize)
names(maxMCIms)
#[1] "idx"     "members"
head(maxMCIms[['idx']])  #  
# 14.5  16.5  18.5 Adult 
# 24     6     5     1 
head(maxMCIms[['members']][['18.5']])
#[1] "0610009B14Rik" "AW495222"      "Abp1"          "Acaa1a"       
#[5] "Acaca"         "Acap3"   

biomodules = getMaxStats(membersL_noweight[['members']], maxMCIms[[1]])
maxMCI = getMaxStats(membersL_noweight[['MCI']],maxMCIms[[1]])
head(maxMCI)
#  14.5      16.5      18.5     Adult 
#22.72876  29.81868 131.83505  14.68497 
CTS.E18 = getCTS(maxMCI, maxMCIms[[2]])
# Length: 205

maxMCI.2nd = getMaxStats(membersL_noweight[['MCI']][-3],maxMCIms[["idx"]][-3])
head(maxMCI.2nd)
#    14.5     16.5    Adult 
# 22.72876 29.81868 14.68497
CTS.E16 = getCTS(maxMCI.2nd, maxMCIms[["members"]][-3])
#Length: 69



# save for repeating and plot (also to Andrew)
save(CTS.E16, CTS.E18, file="outPut.Rdata/GSE52583/Treutlein2014_scRNAseq_lung_CTS_0.05_fdr0.1_minsize30.RData") 



######################################
## estimate significance
########################################

C = 1000
load(file=paste0("GSE52583_GenePermutation_",C,"CTS", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))
simuMCI.E18 <- simulationMCI(length(CTS.E18), samplesL, dat,  B=C, fun='cor')
simuMCI.E18_2nd <- simulationMCI(length(CTS.E18_2nd), samplesL, dat,  B=C)
simuMCI.E16 <- simulationMCI(length(CTS.E16), samplesL, dat,  B=C, fun='cor')
simuMCI.E14<- simulationMCI(length(CTS.E14), samplesL, dat,  B=C, fun='cor')

#save(simuMCI.E14,simuMCI.E16, simuMCI.E18, simuMCI.E18_2nd, file=paste0("GSE52583_GenePermutation_",C,"CTS", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))


#### plot ###################
# load(file=paste0("Shrinked",PCC_gene.target,".Treutlein2014_scRNAseq_lung_CTS_", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))
# load(file=paste0("GSE52583_GenePermutation_",C,"CTS", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))
par(mfrow=c(1,4))
observation = membersL_noweight[["MCI"]][['14.5']]['24']; names(observation) <- '14.5'
plot_MCI_Simulation(observation, simuMCI.E14, ylim=c(0,100),which2point='14.5',las=2,
                    main=paste(length(CTS.E14), "E14-selected genes", "\n","vs. ",C, "times of gene-permutation"))
observation =  membersL_noweight[["MCI"]][['16.5']]['6']; names(observation) <- '16.5'
plot_MCI_Simulation(observation, simuMCI.E16, ylim=c(0,60),which2point='16.5',las=2,
                    main=paste(length(CTS.E16), "E16-selected genes", "\n","vs. ",C, "times of gene-permutation"))
observation =  membersL_noweight[["MCI"]][['18.5']]['5']; names(observation) <- '18.5'      # 441.80687
plot_MCI_Simulation(observation, simuMCI.E18,  ylim=c(0,150), which2point='18.5',las=2,
                    main=paste(length(CTS.E18), "E18-selected genes", "\n","vs. ",C, "times of gene-permutation"))
observation = membersL_noweight[["MCI"]][['18.5']]['1']; names(observation) <- '18.5'
plot_MCI_Simulation(observation, simuMCI.E18_2nd,  ylim=c(0,20), which2point='18.5',las=2,
                    main=paste(length(CTS.E18_2nd), "E18-selected genes", "\n","vs. ",C, "times of gene-permutation"))

dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/MCI_GenePermutation_",C,"CTS", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"))




########################################################################################
####### 3)  Finding Tipping Point and verify using old IC score  ######################
########################################################################################


which2points <- c('16.5','18.5')
CTS <- list('16.5'=CTS.E16, '18.5'=CTS.E18)

for(i in 1:length(CTS)) 
{
  n <- length(CTS[[i]])
  BioTIP_E18 <- getIc(dat, samplesL, genes=CTS[[i]], fun="BioTIP")
  simuBioTIP_g <- simulation_Ic(n, samplesL, dat, B=C,
                                fun="BioTIP")
  simuBioTIP_s <- matrix(nrow=length(samplesL), ncol=C)
  rownames(simuBioTIP_s) = names(samplesL)
  for(j in 1:length(samplesL)) {
    n <- length(samplesL[[j]])  # of cells
    simuBioTIP_s[j,] <- simulation_Ic_sample(dat, sampleNo=n, Ic=BioTIP_E18[which2points[i]],
                                             genes=CTS[[i]],  B=C,
                                             fun="BioTIP")
  }
  
  
  par(mfrow=c(2,2))
  plot_Ic_Simulation(BioTIP_E18, simuBioTIP_g,las = 2, ylab="BioTIP",
                     main=paste(length(CTS[[i]]),names(CTS[i]), "genes", "\n","vs. ",C, "times of gene-permutation"),
                     fun="boxplot")
  plot_SS_Simulation(BioTIP_E18, simuBioTIP_g, 
                     main = paste("Larget - 2nd largest",length(CTS[[i]]),"genes"), #which2point="18.5", 
                     ylab="BioTIP")
  
  plot_Ic_Simulation(BioTIP_E18, simuBioTIP_s,las = 2, ylab="BioTIP",
                     main=paste(length(CTS[[i]]),names(CTS[i]), "genes", "\n","vs. ",C, "times of cell-permutation"),
                     fun="boxplot")
  plot_SS_Simulation(BioTIP_E18, simuBioTIP_s, 
                     main = paste("Larget - 2nd largest",length(CTS[[i]]),"genes"), #which2point="18.5", 
                     ylab="BioTIP")
  
  dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/BioTIP_vsSimulation_",length(CTS[[i]]),"CTS.pdf"))
}




### how about  without shrinkage, the old Ic-score  
for(i in 1:length(CTS)) 
{
  n <- length(CTS[[i]])
  BioTIP_E18 <- getIc(dat, samplesL, genes=CTS[[i]], fun="cor")
  simuBioTIP_g <- simulation_Ic(n, samplesL,dat, B=C,
                                fun="cor")
  simuBioTIP_s <- matrix(nrow=length(samplesL), ncol=C)
  rownames(simuBioTIP_s) = names(samplesL)
  for(j in 1:length(samplesL)) {
    n <- length(samplesL[[j]])  # of cells
    simuBioTIP_s[j,] <- simulation_Ic_sample(dat, sampleNo=n, Ic=BioTIP_E18[which2points[i]],
                                             genes=CTS[[i]],  B=C,
                                             fun="cor")
  }
  
  
  par(mfrow=c(2,2))
  plot_Ic_Simulation(BioTIP_E18, simuBioTIP_g,las = 2, ylab="Ic",
                     main=paste(length(CTS[[i]]),names(CTS[i]), "genes", "\n","vs. ",C, "times of gene-permutation"),
                     fun="boxplot", which2point=which2points[i])
  plot_SS_Simulation(BioTIP_E18, simuBioTIP_g, 
                     main = paste("Larget - 2nd largest",length(CTS[[i]]),"genes"), #which2point="18.5", 
                     ylab="Ic")
  
  plot_Ic_Simulation(BioTIP_E18, simuBioTIP_s,las = 2, ylab="Ic",
                     main=paste(length(CTS[[i]]),names(CTS[i]), "genes", "\n","vs. ",C, "times of cell-permutation"),
                     fun="boxplot", which2point=which2points[i])
  plot_SS_Simulation(BioTIP_E18, simuBioTIP_s, 
                     main = paste("Larget - 2nd largest",length(CTS[[i]]),"genes"), #which2point="18.5", 
                     ylab="Ic")
  
  dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/Ic_vsSimulation_",length(CTS[[i]]),"CTS.pdf"))
}






