myname="AT2"

#### Focus on AT2 branch, 4 states to be studied 

# This code performs CTS analysis on the expressional data of scRNA-seq, using BioTIP (v.1.1.0).
# THe GEO-downloaded FPKM values were normalized by monocle.
# Generated by:  Xinan H Yang (xyang2_at_uchicago.edu)
# Last update: 03/11/2020



library(igraph)
library(BioTIP)
packageVersion("BioTIP")  #[1] '1.1.0'


setwd("F:/projects/BioTIP/doc/2020_/Applications/")
#source("../../../source/BioTIP_update_3.3_02282020.R")


###############################################################
#### read the transcriptome data matrix and sample information
###############################################################  

load(file="F:/projects/BioTIP/doc/2020_/Applications/input.Rdata/GSE52583/GSE52583_monocle_counts.RData")
dim(dat)

load(file="F:/projects/BioTIP/doc/2020_/Applications/input.Rdata/GSE52583/GSE52583_cli.RData")
dim(cli)

# check if sample IDs in the 'cli' matris are identical to the sample names in the 'dat' matrix  
all(rownames(cli) == colnames(dat))  # TRUE

 
###### 1) focus on T2 branch #############
 all(colnames(dat) == rownames(cli))  # TRUE
 ## Method 1  =============
 #x <- which(cli$age=="18.5" & cli$putative_cell_type !="AT2" )
 #length(x)  # 68
 
 ## Method 2  =============
 #y <-  which(cli$age=="18.5" & cli$Cluster =="3" )
 #length(y)   # 52
 
 ## Method 3  =============
 y <-  which(cli$age=="18.5" & cli$CellType !="AT2" )
 length(y)   # 65

 dat <- dat[,-y]; dim(dat)  # [1] 10359    131
 cli <- cli[-y,]; dim(cli)  # 131  21
 table(cli$CellType, cli$age)
 #            14.5 16.5 18.5 Adult
 # Ambiguous    2    0    0     0
 # AT2         10    3    0     0
 # AT2          8   11   15    43
 # Unknown     25   13    0     1
  tmp <- as.vector(cli$age)
  x <- which(tmp=="18.5")
  tmp[x] <- paste(cli$age[x], cli$CellType[x], sep="_")
  table(tmp)
#  14.5     16.5 18.5_AT2    Adult 
#  45       27       15       44 
  tmp <- factor(tmp, levels=c('14.5','16.5','18.5_AT2','Adult'))
  cli$BioTIP_state = tmp
 
  

  dim(dat) # [1] 10359   131
  samplesL <- split(rownames(cli),f = cli$BioTIP_state)
  tmp <- lapply(samplesL, length)
  if(any(tmp<4))  samplesL <- samplesL[-which(tmp<4)] 
  unlist(tmp)
#  14.5     16.5 18.5_AT2    Adult 
#  45       27       15       44 
  

####### 2) find biomarker ######################
# Pre-selection Transcript
cut.preselect = 0.05
cut.fdr = 0.1
cut.minsize = 30

test <- sd_selection(dat, samplesL, cut.preselect)
names(test)
head(test[["16.5"]])

# Network Partition
 
igraphL <- getNetwork(test, fdr = cut.fdr)
# 14.5:517 nodes
# 16.5:517 nodes
# 18.5:521 nodes
# Adult:516 nodes


cluster <- getCluster_methods(igraphL)
names(cluster)
# [1] "14.5"  "16.5"  "18.5_AT2"  "Adult"

cluster[[1]]
# IGRAPH clustering walktrap, groups: 33, mod: 0.69
# + groups:
#   $`1`
# [1] "1700030K09Rik" "Abtb2"         "Arhgef17"      "Arhgef19"      "Arsb"          "Ccdc134"       "Clec16a"      
# [8] "Cxx1a"         "Ddx59"         "Dolk"          "Gna12"         "Gpatch1"       "Heatr2"        "L2hgdh"       
# [15] "Mir692-1"      "Mogat2"        "Myo9b"         "Paqr6"         "Pip5k1c"       "Pld2"          "Ppp4r1l-ps"   
# [22] "Pus7"          "Slc43a1"       "Snora47"       "Srgn"          "Sugp2"         "Tbrg4"         "Tnfrsf21"     
# [29] "Zfp867"        "Zfyve9"       
# 
# $`2`
# [1] "2410021H03Rik" "2610018G03Rik" "Cep135"        "Ephx4"         "Gm20324"       "Gpr39"         "Hdac5"        
# [8] "Hist1h2bb"     "Hist1h2bp"     "Hist3h2ba"     "Icam4"         "Insc"          "Kctd18"        "Klhl28"       
# + ... omitted several groups/vertices


# 2.1) Identifying Dynamic Network Biomodule, old MCI method #########


membersL_noweight <- getMCI(cluster,test, adjust.size = FALSE, fun='cor')
names(membersL_noweight)
#[1] "members" "MCI"     "sd"      "PCC"     "PCCo"  
#plotBar_MCI(membersL_noweight, ylim = c(0,300))
plotBar_MCI(membersL_noweight, ylim = c(0,300), minsize = cut.minsize)
dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/AT2_MCI_bar_preselect", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"))
#the numbers in the parenthesis: they are total number of clusters, no control of the cell (sample) size per  cluster
biomodules = getMaxStats(membersL_noweight[['members']], maxMCIms[[1]])


# Defining CTS

# get the statistics using the old MCI system
maxMCIms <- getMaxMCImember(membersL_noweight[["members"]],membersL_noweight[["MCI"]],min =cut.minsize)
names(maxMCIms)
#[1] "idx"     "members"
head(maxMCIms[['idx']])   
#14.5  16.5  18.5 Adult 
# 24     6     9    15

head(maxMCIms[['members']][['18.5']])
 
maxMCI = getMaxStats(membersL_noweight[['MCI']],maxMCIms[[1]])

CTS.AT2.E18 = getCTS(maxMCI, maxMCIms[[2]])
# Length: 267
MCI.AT2.E18 = max(maxMCI) # 107.39369

save(CTS.AT2.E18, MCI.AT2.E18, file="outPut.Rdata/GSE52583/CTS.AT2.RData")



## estimate MCI significance by boostratp genes NOT repeat !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
 C = 1000
 n <- length(CTS.AT2.E18)
 simuMCI <- simulationMCI(n, samplesL, dat,  B=C, fun="cor")
# save(simuMCI, file=paste0("T2_branch.CellType/GSE52583_GenePermutation_",C,"CTS_AT2.CellType", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))
 
#load(file=paste0("../../../result/GSE52583/T2_branch.CellType/GSE52583_GenePermutation_",C,"CTS_AT2.CellType", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))
 #  simuMCI <- simuMCI[["E18_2nd"]]
 
 names(MCI.AT2.E18) <- "18.5"
 plot_MCI_Simulation(MCI.AT2.E18, simuMCI, las=2,ylim=c(0,110),
                     main=paste("E16.5 CTS", n, "genes",
                                "\n","vs. ",C, "times of gene-permutation"), which2point="18.5")
 dev.copy2pdf(file="ROutputFigs/GSE52583/MCI_GenePermutation_1000CTS_AT2.CellType0.05_fdr0.1_minsize30.pdf")
 
## END of NOT repeat ############################ 


 
 ######## 3)  Finding Tipping Point and verify using IC score  #################
 
 ##################################################
 ######  BioTIP score, shulffing genes ##############
 C= 1000
 CTS <- CTS.AT2.E18
 n <- length(CTS)
 BioTIP_score <- getIc(dat, samplesL, CTS, fun="BioTIP", shrink=TRUE)
 BioTIP_score
 #        14.5        16.5    18.5_AT2       Adult 
 #   0.001826992 0.006089339 1.270759994 0.001503414 
 
 simuBioTIP_g  <- simulation_Ic(n, samplesL, dat, B=C,
                                fun="BioTIP", shrink=TRUE)
 
 #### BioTIP score,  shulffing labelling
 x <- 3
 #  ns <- length(samplesL[[x]])  # of cells at the state of interest
 simuBioTIP_s <- matrix(nrow=length(samplesL), ncol=C)
 rownames(simuBioTIP_s) = names(samplesL)
 for(j in 1:length(samplesL)) {
   ns <- length(samplesL[[j]])  # of each state
   simuBioTIP_s[j,] <- simulation_Ic_sample(dat, ns, Ic=BioTIP_score[x],
                                            genes=CTS, B=C,
                                            fun="BioTIP", shrink=TRUE)
 }
# save( simuBioTIP_g,  simuBioTIP_s, file="F:/projects/BioTIP/result/GSE52583/AT2_simuBioTIP.RData", compress=TRUE)
 
 # load(file="F:/projects/BioTIP/result/GSE52583/AT2_simuBioTIP.RData")  
 pdf(file="ROutputFigs/GSE52583/AT2.matplot_density_Ic_simulation.pdf",width=10, height=7)  
 par(mfrow=c(1,2))
 ## global matplot 
 plot_Ic_Simulation(BioTIP_score, simuBioTIP_g, las = 2, ylab="BioTIP",
                    main= paste("CTS.E18 (",n," transcripts)"),
                    fun="matplot", which2point= '18.5_AT2')
 ## #local Kernel Density Plot  
 d <- density(simuBioTIP_s['18.5_AT2',], ylim=c(0,1.2)) # returns the density data
 plot(d) # plots the results
 abline(v=BioTIP_score['18.5_AT2'], col="green")
 dev.off() 
 
 
 ## supporting GitHub plots ###      
 par(mfrow=c(2,2))
 
 plot_Ic_Simulation(BioTIP_score, simuBioTIP_g, las = 2, ylab="BioTIP", ylim=c(0,0.12),
                    main=paste(length(CTS),names(samplesL)[x], "genes", "\n","vs. ",C, "gene-permutations"),
                    fun="boxplot", which2point= x)
 plot_SS_Simulation(BioTIP_score, simuBioTIP_g, 
                    main = paste("Delta BioTIP",n,"genes"), 
                    ylab=NULL)
 plot_Ic_Simulation(BioTIP_score, simuBioTIP_s, las = 2, ylab="BioTIP", ylim=c(0, 1.2),
                    main=paste(length(CTS), names(samplesL)[x],"genes", "\n","vs. ",C, "sample-permutations"),
                    fun="matplot", which2point= x)
 plot_SS_Simulation(BioTIP_score, simuBioTIP_s, 
                    main = "Delta BioTIP, label shuffling", 
                    ylab=NULL)
 
 dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/AT2_BioTIP_vsSimulation_",length(CTS),"CTS.pdf"))
 wilcox.test(simuBioTIP_g[1,], simuBioTIP_g[2,]) # p-value < 2.2e-16
 wilcox.test(simuBioTIP_g[3,], simuBioTIP_g[2,]) # p-value < 2.2e-16
 
 
 ##################################################
 ######  Ic score, of the same number of random genes ##############
 C= 1000
 n <- length(CTS)
 Ic_score <- getIc(dat, samplesL, CTS, fun="cor")
 Ic_score
#    14.5      16.5   18.5_AT2     Adult 
# 0.1582122 0.8776503 3.8005775 0.3883955 
 
 simuIc_g  <- simulation_Ic(n, samplesL, dat, B=C,
                            fun='cor')
 
 
 #### Ic score for the identified CTS,  shulffing labelling
 x <- 2
 #  ns <- length(samplesL[[x]])  # of fixed number of cells
 simuIc_s <- matrix(nrow=length(samplesL), ncol=C)
 rownames(simuIc_s) = names(samplesL)
 for(j in 1:length(samplesL)) { 
   ns <- length(samplesL[[j]])  # for each state rewpectively 
   simuIc_s[j,] <- simulation_Ic_sample(dat, ns, Ic=Ic_score[x],
                                        genes=CTS, B=C,
                                        fun="cor")
 }
# save( simuIc_g,  simuIc_s, file="F:/projects/BioTIP/result/GSE52583/AT2_simuIC.RData", compress=TRUE)
 
 # load(file="F:/projects/BioTIP/result/GSE52583/AT2_simuIC.RData")  
 
 x <- 3
 par(mfrow=c(2,2))
 
 plot_Ic_Simulation(Ic_score, simuIc_g, las = 2, ylab="Ic",
                    main=paste(length(CTS),names(samplesL)[x], "genes", "\n","vs. ",C, "gene-permutations"),
                    fun="boxplot", which2point= x)
 plot_SS_Simulation(Ic_score, simuIc_g, 
                    main = paste("Delta Ic",n,"genes"), 
                    ylab=NULL)
 plot_Ic_Simulation(Ic_score, simuIc_s, las = 2, ylab="Ic", ylim=c(0, 6),
                    main=paste(length(CTS), names(samplesL)[x],"genes", "\n","vs. ",C, "sample-permutations"),
                    fun="matplot", which2point= x)
 plot_SS_Simulation(Ic_score, simuIc_s, 
                    main = "Delta BioTIP, label shuffling", 
                    ylab=NULL)
 
 dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/AT2_Ic_vsSimulation_",length(CTS),"CTS.pdf"))
 wilcox.test(simuIc_g[1,], simuIc_g[2,]) # p-value < 2.2e-16
 wilcox.test(simuIc_g[3,], simuIc_g[2,]) # p-value < 2.2e-16
 
 
 
 
 ######## 3.2)  Whetehr the E16 finding (70 transcripts) is statistically significant   #################
 
 ##################################################
 ######  BioTIP score, shulffing genes ##############
 C= 1000
 load('outPut.Rdata/GSE52583/CTS_AT2.E16.5.RData')
 CTS <- CTS.AT2.E16
 (n <- length(CTS))  # 70
 BioTIP_score <- getIc(dat, samplesL, CTS, fun="BioTIP", shrink=TRUE)
 BioTIP_score
 #      14.5       16.5   18.5_AT2      Adult 
 #  0.11045763 1.14451424 0.18605761 0.01597367 
 
   
 simuBioTIP_g  <- simulation_Ic(n, samplesL, dat, B=C,
                                fun="BioTIP", shrink=TRUE)
 
 #### BioTIP score,  shulffing labelling
 x <- 2
 #  ns <- length(samplesL[[x]])  # of cells at the state of interest
 simuBioTIP_s <- matrix(nrow=length(samplesL), ncol=C)
 rownames(simuBioTIP_s) = names(samplesL)
 for(j in 1:length(samplesL)) {
   ns <- length(samplesL[[j]])  # of each state
   simuBioTIP_s[j,] <- simulation_Ic_sample(dat, ns, Ic=BioTIP_score[x],
                                            genes=CTS, B=C,
                                            fun="BioTIP", shrink=TRUE)
 }
 # save( simuBioTIP_g,  simuBioTIP_s, file="F:/projects/BioTIP/result/GSE52583/AT2_simuBioTIP_E16.RData", compress=TRUE)
 
 # load(file="F:/projects/BioTIP/result/GSE52583/AT2_simuBioTIP_E16.RData")  
 pdf(file="ROutputFigs/GSE52583/AT2.E16.5.matplot_density_Ic_simulation.pdf",width=10, height=7)  
 par(mfrow=c(1,2))
 ## global matplot 
 plot_Ic_Simulation(BioTIP_score, simuBioTIP_g, las = 2, ylab="BioTIP",
                    main= paste("CTS.E16 (",n," transcripts)"),
                    fun="boxplot", which2point= '16.5')
 ## #local Kernel Density Plot  
 d <- density(simuBioTIP_s['16.5',]) # returns the density data
 plot(d) # plots the results
 abline(v=BioTIP_score['16.5_AT2'], col="green")
 dev.off() 
 
 
 ## supporting GitHub plots ###      
 par(mfrow=c(2,2))
 
 plot_Ic_Simulation(BioTIP_score, simuBioTIP_g, las = 2, ylab="BioTIP", ylim=c(0,1),
                    main=paste(length(CTS),names(samplesL)[x], "genes", "\n","vs. ",C, "gene-permutations"),
                    fun="boxplot", which2point= x)
 plot_SS_Simulation(BioTIP_score, simuBioTIP_g, 
                    main = paste("Delta BioTIP",n,"genes"), 
                    ylab=NULL)
 plot_Ic_Simulation(BioTIP_score, simuBioTIP_s, las = 2, ylab="BioTIP", ylim=c(0, 10),
                    main=paste(length(CTS), names(samplesL)[x],"genes", "\n","vs. ",C, "sample-permutations"),
                    fun="matplot", which2point= x)
 plot_SS_Simulation(BioTIP_score, simuBioTIP_s, 
                    main = "Delta BioTIP, label shuffling", 
                    ylab=NULL)
 
 dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/AT2.E16.5_BioTIP_vsSimulation_",length(CTS),"CTS.pdf"))
 
 
 
 
 #############################################
 ### 4) motif analysis , do not repeat ####################### 
   
 # # https://www.bioconductor.org/packages/release/workflows/vignettes/generegulation/inst/doc/generegulation.html
 # library(Biostrings)
 # library(GenomicFeatures)
 # 
 # load(file="GSE52583_annot.cds_Clustered_thresholded.RData")
 # dim(annot.cds) # [1] 22854   196
 # 
 # G <- fData(annot.cds)
 # dim(G) #[1] 22854    11
 # colnames(G)
 # # [1] "ensembl_gene_id"     "gene_short_name"     "chromosome_name"     "strand"             
 # # [5] "start_position"      "end_position"        "gene_biotype"        "GSE52583.gene.id"   
 # # [9] "locus"               "num_cells_expressed" "use_for_ordering"   
 # 
 # annot.cds <-annot.cds[G$use_for_ordering,]
 # 
 # dat <- exprs(annot.cds)
 # dim(dat) #dim(dat) # [1] 10359   196
 # G <- subset(G, use_for_ordering==TRUE)
 # dim(G)  #[1] 10359    11
 # 
 # 
 # 
 #   resPath = 'FASTA_TSS_1500_500'
 #   upstream=1500
 #   downstream=500
 #   GS <- CTS.AT2.E18
 # 
 #   x <- intersect(GS, rownames(dat))
 #   cat(paste(length(x), 'out of', length(GS),'genes of interest are expressed', '\t'))
 #   chromosomal.loc <- GRanges(G[x,]$locus)
 #   names(chromosomal.loc) <- rownames(G[x,])
 #     
 #     ## get promoter loci and write into bed file
 #   promoters <- promoters(chromosomal.loc, upstream=upstream, downstream=downstream)
 #   # remove blacklist
 #   blacklist.mm10 <- import('F:\\projects\\Share\\Blacklisted\\2019\\mm10-blacklist.v2.bed')
 #   blacklisted <- findOverlaps(promoters, blacklist.mm10, type="within")
 #   cat(length(blacklisted),'in blacklist', '\n')
 #   if(length(blacklisted)>0)     promoters <- promoters[-from(blacklisted)]
 #   export.bed(promoters, con="outPut.Rdata/GSE52583/CTS_AT2.E18.5.bed")
 #   # 267 out of 267 genes of interest are expressed 4 in blacklist
     

 #Homer findMotifs.pl  CTS_AT2.E18.5.bed mm10 GSE115575_1500_500/ 
 

 