myname="AT1"

#### Focus on AT1 branch, 3 states to be studied 

# This code performs CTS analysis on the expressional data of scRNA-seq, using BioTIP (v.1.1.1).
# The GEO-downloaded FPKM values were normalized by monocle.
# Generated by:  Xinan H Yang (xyang2_at_uchicago.edu)
# Last update: 02/25/2021



library(igraph)

# To install the newest BioTIP version, you can do:
# library("devtools")
# devtools::install_github("xyang2uchicago/BioTIP")

library(BioTIP)
package.version('BioTIP') # [1] "1.5.0"

# Set up your working directory, which could be customized
setwd("F:/projects/BioTIP/doc/2020_/Applications/")
#source("../../../source/BioTIP_update_3.3_02282020.R")


###############################################################
#### read the transcriptome data matrix and sample information
###############################################################  

load(file="F:/projects/BioTIP/doc/2020_/Applications/input.Rdata/GSE52583/GSE52583_monocle_counts.RData")
dim(dat)

load(file="F:/projects/BioTIP/doc/2020_/Applications/input.Rdata/GSE52583/GSE52583_cli.RData")
dim(cli)

# check if sample IDs in the 'cli' matris are identical to the sample names in the 'dat' matrix  
all(rownames(cli) == colnames(dat))  # TRUE
#any(rownames(cli) != colnames(dat))  # FALSE


 
###### 1)  focus on T1 branch #############
 y <-  which((cli$age=="18.5" & (cli$CellType !="AT1"   | cli$cell_type != "18.5_AT1") )
             | cli$age=="Adult" ) 
 length(y)   # 86   
 
 dat <- dat[,-y]; dim(dat)  # [1] 10359    110
 cli <- cli[-y,]; dim(cli)  # 110  21
 table(cli$CellType, cli$age)
 #            14.5 16.5 18.5 Adult
 # Ambiguous    2    0    0     0
 # AT1         10    3    38    0
 # AT2          8   11    0    0
 # Unknown     25   13    0     0
 tmp <- as.vector(cli$age)
 x <- which(tmp=="18.5")
 tmp[x] <- paste(cli$age[x], cli$CellType[x], sep="_")
 table(tmp)
 #  14.5     16.5 18.5_AT1     
 #  45       27       38         
 tmp <- factor(tmp, levels=c('14.5','16.5','18.5_AT1'))
 cli$BioTIP_state = tmp
 
 
 
 dim(dat) # [1] 10359   110
 samplesL <- split(rownames(cli),f = cli$BioTIP_state)
 (tmp <- lapply(samplesL, length))
 if(any(tmp<4))  samplesL <- samplesL[-which(tmp<4)] 
 
 

####### 2) find gene modules ######################
# Pre-selection Transcript
cut.preselect = 0.05
cut.fdr = 0.1
cut.minsize = 30

## The current function works on matrix but not "dgCMatrix"
if(class(dat)=="dgCMatrix") dat <- as.matrix(dat)

test <- sd_selection(dat, samplesL, cut.preselect)
names(test)
head(test[["16.5"]])
# if the output is a numeric, you might wrongly use dgCMatrix for the sd_selection()
class(test[[1]])
# [1] "matrix" "array" 


# Network Partition
 
igraphL <- getNetwork(test, fdr = cut.fdr)
# 14.5:517 nodes
# 16.5:517 nodes
# 18.5:516 nodes


cluster <- getCluster_methods(igraphL)
names(cluster)
# [1] "14.5"  "16.5"  "18.5_AT1" 


# 2.1) Identifying CTS, using MCI (i.e. DNB) method #########

membersL_noweight <- getMCI(cluster,test, adjust.size = FALSE)

names(membersL_noweight)
#[1] "members" "MCI"     "sd"      "PCC"     "PCCo"  
pdf(file=paste0("ROutputFigs/GSE52583/AT1_MCI_bar_preselect", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"),
    width=10, height=3)
#plotBar_MCI(membersL_noweight, ylim = c(0,300))
plotBar_MCI(membersL_noweight, ylim = c(0,50), minsize = cut.minsize)
dev.off
# Note that in the output bar plot, the numbers on top are the total number of modules (gene clusters) per state.
# Showing the bars with more than the 'minsize' of genes, with the number of genes on the top of that bar.
# X-axis is the module ID within each state which may be truncated due to plot-width.

# Get the statistics using the MCI system, for the most top module per state
maxMCI.per.state <- getMaxMCImember(membersL_noweight[["members"]],membersL_noweight[["MCI"]], 
                            min =cut.minsize, n=1)

names(maxMCI.per.state)
#[1] "idx"     "members"
head(maxMCI.per.state[['idx']])   
# 14.5     16.5     18.5_AT1  
#  21        4        2

# When set n=2 in this function, the output will have one more object "2topest.members", 
# and two module IDs per state in the 'maxMCI.per.state$idx' object (when existing).
# In this case, you can extract the MCI score down to the 2nd highest bar in a state pf interest,
# by calling function 
maxMCI.per.state_2 <- getMaxMCImember(membersL_noweight[["members"]],membersL_noweight[["MCI"]], 
                            min =cut.minsize, n=2
names(maxMCI.per.state)
#[1] "idx"     "members"    "2topest.members"
head(maxMCI.per.state[['idx']][['16.5']])  
# 16.5     
# 4  21                                      
getNextMaxStats(membersL_noweight[['MCI']], idL = maxMCI.per.state_2[['idx']], 
                                whoisnext='16.5', which.next = 2)
# 

## Extract candidate module genes
head(maxMCI.per.state[['members']][['18.5_AT1']])
 
maxMCI = getMaxStats(membersL_noweight[['MCI']], maxMCI.per.state[['idx']])
head(maxMCI)
#   14.5      16.5      18.5_AT1      
# 10.01716   28.77452   18.88909

# Extract the CTS genes for the state with the highest MCI in the system,
# which is the predicted 'tipping point.'
CTS.AT1.E16 =  getCTS(maxMCI, membersL_noweight[["members"]])
# Length: 71

## Extract n states of which the highest MCI per are the top score in the system.
## This function is important when analyzing a complex system with multiple tipping points.
getTopMCI(membersL_noweight[["members"]],  membersL_noweight[["MCI"]],  
                       membersL_noweight[["MCI"]], 
                       min = cut.minsize, n = 2)  
# 16.5      18.5_AT1      
# 28.77452   18.88909                                      
## Here, we set n=1 to focus on the state '16.5', i.e., the state with the most top MCI.
MCI.AT1.E16 = getTopMCI(membersL_noweight[["members"]],  membersL_noweight[["MCI"]],  
                       membersL_noweight[["MCI"]], 
                       min = cut.minsize, n = 1)
# Alternatively, in this example, we can simply get the score by
max(maxMCI) # 28.77452

## In case you want to look into more modules,
CTS.AT1.E16.2nd <- maxMCI.per.state_2[['2topest.members']]['16.5'])                                       
length(CTS.AT1.E16.2nd)  #38   

getNextMaxStats(membersL_noweight[['MCI']], idL = maxMCI.per.state_2[['idx']], 
                                whoisnext='16.5', which.next = 2)
                                      
save(CTS.AT1.E16, MCI.AT1.E16, file="outPut.Rdata/GSE52583/CTS.AT1.RData")
 


## Estimate significance, NOT repeat (takes a while)
 C = 1000
 n <- length(CTS.AT1.E16)
 simuMCI <- simulationMCI(n, samplesL, dat,  B=C)
 
#save(simuMCI, file=paste0("T1_branch.CellType/GSE52583_GenePermutation_",C,"CTS_AT1.CellType", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))

# load("../../../result/GSE52583/T1_branch.CellType/GSE52583_GenePermutation_1000CTS_AT1.CellType0.05_fdr0.1_minsize30.RData")
#  simuMCI <- simuMCI[["E16_1st"]]
 names(MCI.AT1.E16) <- "16.5"
plot_MCI_Simulation(MCI.AT1.E16, simuMCI, las=2,ylim=c(0,30),
                    main=paste("E16.5 CTS", n, "genes",
                               "\n","vs. ",C, "times of gene-permutation"), which2point="16.5")
dev.copy2pdf(file="ROutputFigs/GSE52583/MCI_GenePermutation_1000CTS_AT1.CellType0.05_fdr0.1_minsize30.pdf")



######## 3)  Find and verify Tipping Point using IC* score  #################

##################################################
######  BioTIP score, shulffing genes ##############
C= 1000
CTS <- CTS.AT1.E16
n <- length(CTS)
BioTIP_score <- getIc(dat, samplesL, CTS, fun="BioTIP", shrink=TRUE, PCC_sample.target='average')
BioTIP_score
#      14.5       16.5   18.5_AT1 
#  0.11850833 1.29987843 0.06653507

simuBioTIP_g  <- simulation_Ic(n, samplesL, dat, B=C,
                               fun="BioTIP", shrink=TRUE)

#### Ic* score,  shulffing labelling
x <- 2
#  ns <- length(samplesL[[x]])  # of cells at the state of interest
simuBioTIP_s <- matrix(nrow=length(samplesL), ncol=C)
rownames(simuBioTIP_s) = names(samplesL)
for(j in 1:length(samplesL)) {
    ns <- length(samplesL[[j]])  # of each state
    simuBioTIP_s[j,] <- simulation_Ic_sample(dat, ns, Ic=BioTIP_score[x],
                                             genes=CTS, B=C,
                                             fun="BioTIP", shrink=TRUE)
}
# save( simuBioTIP_g,  simuBioTIP_s, file="F:/projects/BioTIP/result/GSE52583/AT1_simuBioTIP.RData", compress=TRUE)

# load(file="F:/projects/BioTIP/result/GSE52583/AT1_simuBioTIP.RData")  
pdf(file="ROutputFigs/GSE52583/AT1.matplot_density_Ic_simulation.pdf",width=10, height=7)  
par(mfrow=c(1,2))
## global matplot 
plot_Ic_Simulation(BioTIP_score, simuBioTIP_g, las = 2, ylab="BioTIP",
                   main= paste("CTS.E16 (",n," transcripts)"),
                   fun="matplot", which2point= '16.5')
## #local Kernel Density Plot  
d <- density(simuBioTIP_s['16.5',]) # returns the density data
plot(d) # plots the results
abline(v=BioTIP_score['16.5'], col="green")
dev.off() 


## Supporting GitHub plots ###      
par(mfrow=c(2,2))

plot_Ic_Simulation(BioTIP_score, simuBioTIP_g, las = 2, ylab="BioTIP", ylim=c(0,0.12),
                   main=paste(length(CTS),names(samplesL)[x], "genes", "\n","vs. ",C, "gene-permutations"),
                   fun="boxplot", which2point= x)
plot_SS_Simulation(BioTIP_score, simuBioTIP_g, 
                   main = paste("Delta BioTIP",n,"genes"), 
                   ylab=NULL)
plot_Ic_Simulation(BioTIP_score, simuBioTIP_s, las = 2, ylab="BioTIP", ylim=c(0, 4),
                   main=paste(length(CTS), names(samplesL)[x],"genes", "\n","vs. ",C, "sample-permutations"),
                   fun="matplot", which2point= x)
plot_SS_Simulation(BioTIP_score, simuBioTIP_s, 
                   main = "Delta BioTIP, label shuffling", 
                   ylab=NULL)

dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/AT1_BioTIP_vsSimulation_",length(CTS),"CTS.pdf"))
wilcox.test(simuBioTIP_g[1,], simuBioTIP_g[2,]) # p-value < 2.2e-16
wilcox.test(simuBioTIP_g[3,], simuBioTIP_g[2,]) # p-value < 2.2e-16


##################################################
###### The previous Ic score, of the same number of random genes ##############
C= 1000
n <- length(CTS)
Ic_score <- getIc(dat, samplesL, CTS, fun="cor", PCC_sample.target='average')
Ic_score
#    14.5     16.5 18.5_AT1 
# 1.076046 5.775612 2.280395 
simuIc_g  <- simulation_Ic(n, samplesL, dat, B=C,
                           fun='cor')


#### The previous Ic score for the identified CTS,  shulffing labelling
x <- 2
#  ns <- length(samplesL[[x]])  # of fixed number of cells
simuIc_s <- matrix(nrow=length(samplesL), ncol=C)
rownames(simuIc_s) = names(samplesL)
for(j in 1:length(samplesL)) { 
    ns <- length(samplesL[[j]])  # for each state rewpectively 
    simuIc_s[j,] <- simulation_Ic_sample(dat, ns, Ic=Ic_score[x],
                                         genes=CTS, B=C,
                                         fun="cor")
}
# save( simuIc_g,  simuIc_s, file="F:/projects/BioTIP/result/GSE52583/AT1_simuIC.RData", compress=TRUE)

# load(file="F:/projects/BioTIP/result/GSE52583/AT1_simuIC.RData")  


par(mfrow=c(2,2))

plot_Ic_Simulation(Ic_score, simuIc_g, las = 2, ylab="Ic",
                   main=paste(length(CTS),names(samplesL)[x], "genes", "\n","vs. ",C, "gene-permutations"),
                   fun="boxplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_g, 
                   main = paste("Delta Ic",n,"genes"), 
                   ylab=NULL)
plot_Ic_Simulation(Ic_score, simuIc_s, las = 2, ylab="Ic", ylim=c(1, 10),
                   main=paste(length(CTS), names(samplesL)[x],"genes", "\n","vs. ",C, "sample-permutations"),
                   fun="matplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_s, 
                   main = "Delta BioTIP, label shuffling", 
                   ylab=NULL)

dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/AT1_Ic_vsSimulation_",length(CTS),"CTS.pdf"))
wilcox.test(simuIc_g[1,], simuIc_g[2,]) # p-value < 2.2e-16
wilcox.test(simuIc_g[3,], simuIc_g[2,]) # p-value < 2.2e-16


#############################################
### 3) motif analysis , do not repeat ####################### 

# # https://www.bioconductor.org/packages/release/workflows/vignettes/generegulation/inst/doc/generegulation.html
# library(Biostrings)
# library(GenomicFeatures)
# 
# load(file="GSE52583_annot.cds_Clustered_thresholded.RData")
# dim(annot.cds) # [1] 22854   196
# 
# G <- fData(annot.cds)
# dim(G) #[1] 22854    11
# colnames(G)
# # [1] "ensembl_gene_id"     "gene_short_name"     "chromosome_name"     "strand"             
# # [5] "start_position"      "end_position"        "gene_biotype"        "GSE52583.gene.id"   
# # [9] "locus"               "num_cells_expressed" "use_for_ordering"   
# 
# annot.cds <-annot.cds[G$use_for_ordering,]
# 
# dat <- exprs(annot.cds)
# dim(dat) #dim(dat) # [1] 10359   196
# G <- subset(G, use_for_ordering==TRUE)
# dim(G)  #[1] 10359    11
# 
# 
# 
#   resPath = 'FASTA_TSS_1500_500'
#   upstream=1500
#   downstream=500
#   GS <- CTS.AT1.E16
# 
#   x <- intersect(GS, rownames(dat))
#   cat(paste(length(x), 'out of', length(GS),'genes of interest are expressed', '\t'))
#   chromosomal.loc <- GRanges(G[x,]$locus)
#   names(chromosomal.loc) <- rownames(G[x,])
#     
#     ## get promoter loci and write into bed file
#   promoters <- promoters(chromosomal.loc, upstream=upstream, downstream=downstream)
#   # remove blacklist
#   blacklist.mm10 <- import('F:\\projects\\Share\\Blacklisted\\2019\\mm10-blacklist.v2.bed')
#   blacklisted <- findOverlaps(promoters, blacklist.mm10, type="within")
#   cat(length(blacklisted),'in blacklist', '\n')
#   if(length(blacklisted)>0)     promoters <- promoters[-from(blacklisted)]
#   export.bed(promoters, con="outPut.Rdata/GSE52583/CTS_AT2.E18.5.bed")
#   # 71 out of 71 genes of interest are expressed 3 in blacklist   

# Homer findMotifs.pl  CTS_AT1.E16.5.bed mm10 GSE115575_1500_500/ 

