# This code exaime the Tbox- and Gli-bound instances among the identified CTS promoters.
# Generated by:  Xinan H Yang (xyang2_at_uchicago.edu)
# Last update: 03/20/2020

setwd("F:/projects/BioTIP/doc/2020_/Applications/")

library(GenomicRanges)
library(rtracklayer)

#Homer: findMotifs.pl  promoter_1500_500_noBlacklist.bed mm10 GSE52583/homer/TSS_1500_500/
#Homer: scanMotifGenomeWide.pl /gpfs/data/yang-lab/projects/GSE52583/homer/TSS_1500_500/CTS_AT1.E16.5.bed/knownResults/known2.motif mm10 -bed > /gpfs/data/yang-lab/projects/GSE52583/homer/Tbx5.mm10.instance.bed
#Homer: scanMotifGenomeWide.pl  scanMotifGenomeWide.pl /gpfs/data/yang-lab/projects/GSE52583/homer/TSS_1500_500/CTS_AT1.E16.5.bed/knownResults/known2.motif mm10 -bed > /gpfs/data/yang-lab/projects/GSE52583/homer/Tbx5.mm10.instance.bed



### find the instance among CTS promotes
Glis <- import('F:/projects/BioTIP/result/GSE52583/homer/Glis.mm10.instance.bed')
length(Glis)   #[1] 15200
Tbox <- import('F:/projects/BioTIP/result/GSE52583/homer/Tbx5.mm10.instance.bed')
length(Tbox)   #[1] 4864851

## load the identified CTS
load('outPut.Rdata/GSE52583/CTS.AT1.RData')
length(CTS.AT1.E16)  # 71

load('outPut.Rdata/GSE52583/CTS_AT2.E16.5.RData')
length(CTS.AT2.E16)  # 70

load('outPut.Rdata/GSE52583/Treutlein2014_scRNAseq_lung_CTS_0.05_fdr0.1_minsize30.RData')
length(CTS.E16)  # 69
length(CTS.E18)  # 205

## load the black-list masked CTS promoters
AT1.E16.promoter <- import('outPut.Rdata/GSE52583/CTS_AT1.E16.5.bed')
length(AT1.E16.promoter)  #[1] 68
AT2.E16.promoter <- import('outPut.Rdata/GSE52583/CTS_AT2.E16.5.bed')
length(AT2.E16.promoter)  #[1] 67

# all.E18.promoter <- import('outPut.Rdata/GSE52583/CTS_all.E18.5.bed')
# length(all.E18.promoter)  #[1] 202
# AT2.E18.promoter <- import('outPut.Rdata/GSE52583/CTS_AT2.E18.5.bed')
# length(AT2.E18.promoter)   #[1] 263


AT1.Tbox <- AT1.E16.promoter[queryHits(findOverlaps(AT1.E16.promoter, Tbox)),]
setdiff(CTS.AT1.E16,  unique(AT1.Tbox$name))
# [1] "1810043G02Rik" "2610008E11Rik" "Acta2"         "C8g"          
# [5] "Clcn4-2"       "Csf2ra"        "Lrp3"          "Mtrr"         
# [9] "Parm1"         "Slc41a3"       "Sult5a1"       "Tbx5"  

AT2.Tbox <- AT2.E16.promoter[queryHits(findOverlaps(AT2.E16.promoter, Tbox)),]
setdiff(CTS.AT2.E16, unique(AT2.Tbox$name))
# [1] "1810043G02Rik" "2610008E11Rik" "Acta2"         "C8g"          
# [5] "Csf2ra"        "Fhl1"          "Lrp3"          "Mtrr"         
# [9] "Sfxn3"         "Slc41a3"       "Sult5a1"       "Tbx5" 


#samplesL[['18.5_AT2']] <-  rownames(cli)[which(cli$age=="18.5" & cli$CellType =="AT2" )]
# samplesL[['18.5_AT1']] <-  rownames(cli)[which(cli$age=="18.5" & cli$CellType =="AT1" )]
# samplesL[['18.5_other']] <- setdiff(samplesL[['18.5']], samplesL[['18.5_AT1']])
# samplesL[['18.5_other']] <- setdiff(samplesL[['18.5_other']], samplesL[['18.5_AT2']])
# samplesL <- samplesL[c("14.5",  "16.5",  "18.5_AT1", "18.5_AT2", "18.5_other","Adult")]


load(file="F:/projects/BioTIP/doc/2020_/Applications/input.Rdata/GSE52583/GSE52583_monocle_counts.RData")
dim(dat)

load(file="F:/projects/BioTIP/doc/2020_/Applications/input.Rdata/GSE52583/GSE52583_cli.RData")
dim(cli)

all(rownames(cli) != colnames(dat))  # TRUE

samplesL <- split(rownames(cli),f = cli$age)
tmp <- lapply(samplesL, length)
if(any(tmp<4))  samplesL <- samplesL[-which(tmp<4)] 


quicklines=function( stageL, genes, intG, highlightG, stagev2, main){
 # stageL = stageL[grep(stage, names(stageL))]
  samplen = sapply(stageL, ncol)
  idx = which(names(samplen) == stagev2)
  idx1 = ifelse(idx ==1, 0, sum(samplen[1:(idx-1)])-1)
  idx2 = sum(samplen[1:idx])+1
  tmp = do.call(cbind, stageL)  
  toplot = subset(tmp, row.names(tmp) %in% genes)
  toplot = toplot - apply(toplot,1,mean)
 # biotype = full_table$type.toPlot[match(row.names(tmp),full_table$Row.names)]
  color = ifelse(rownames(toplot) %in% intG,'grey','blue')
  if(any(rownames(toplot)  %in% highlightG)) color[which(rownames(toplot)  %in% highlightG)] <- 'red'
  matplot(t(toplot),type = 'l',main = main, lty =1, col = color)
  abline(v = idx1, col = 'orange',lty = 2)
  abline(v = idx2, col = 'orange',lty = 2)
}

#datL = lapply(samplesL, function(x) log2(dat+1)[,x])
datL = lapply(samplesL, function(x) dat[,x])
names(datL) = names(samplesL)

par(mfrow=c(1,2))
quicklines(datL, CTS.AT1.E16, unique(AT1.Tbox$name), highlightG="Tbx5",  stagev2="16.5", main="AT1.E16.CTS")
quicklines(datL, CTS.AT2.E16, unique(AT2.Tbox$name), highlightG="Tbx5",  stagev2="16.5", main="AT2.E16.CTS")
#dev.copy2pdf(file="ROutputFigs/GSE52583/metaplot_AT1.E16.CTS_Tbx5_log2.pdf")
dev.copy2pdf(file="ROutputFigs/GSE52583/metaplot_AT1.E16.CTS_Tbx5.pdf")


##### Functional analysis of the TBC5-targets among the E16_CTS genes
write.table(unique(AT1.Tbox$name), file='outPut.Rdata/GSE52583/Tbox_CTS_AT1.E16.5.txt',sep="\t", row.names=FALSE)
length(unique(AT1.Tbox$name))  # 59

# IPA annotation recognized 55 out of the 59 Tbox-targets
# IPA run core analysis on Tbx5 and these 55 genes 
