# This code performs motif analysi on the BioTIP-identified CTSs from data of GSE115575.
# Generated by: Xinan H Yang
# Revised by: Xinan H Yang (xyang2_at_uchicago.edu)
# Last update: 03/06/2020

setwd("F:/projects/BioTIP/doc/2020_/Applications/")



library(GenomicRanges)


############# load the identification by Antonio ########################

load('outPut.Rdata/GSE115575/weighted_cyclic_loess_signature.RData')
class(weighted_cyclic_loess_signature)  #[1] "data.frame"
dim(weighted_cyclic_loess_signature)  #[1] 253   6
head(weighted_cyclic_loess_signature)
tmp <- weighted_cyclic_loess_signature
loci <- IRanges(start=tmp$start,end=tmp$end,names=tmp$name)
mygr <-  GRanges(seqnames=Rle(tmp$chrom),ranges=loci, strand=tmp$strand)
mygr  #GRanges object with 253 ranges and 0 metadata columns:
any(duplicated(names(mygr))) #[1] TRUE

mypromoter <- promoters(mygr, 500, 200)

### defind the promoters and write bed files for HOMER2 running (USED!!) 

## These objects are made from the so-called black list or signal artefact regions defined by Anshul Kundaje Alan Boyle as part of the ENCODE and modENCODE projects. 
## They are regions which show a signal in essentially any ChIP or similar experiment. For this reason it is useful to remove reads aligning to these regions 
## before carrying out any sort of functional genomics analysis.
library(rtracklayer)
tmp <- read.table(file='F:/projects/Share/Blacklisted/2019/hg38-blacklist.v2.bed', sep='\t',header=F)
dim(tmp)
#[1] 636   4
table(tmp[,4])
#High Signal Region    Low Mappability 
#               594                 42
BK <- GRanges(seqnames=Rle(tmp[,1]),ranges=IRanges(start=tmp[,2],end=tmp[,3]))

 RemoveInterval = function(x, y) {
 ## output the remaining interval of x after removing its overlapInterval with y 
     rv <- x
     start(rv) <- 0
     end(rv) <- 1
     if(start(x) < start(y)) {
            rv = x
            start(rv) = start(x)
            end(rv) = start(y)
            if (end(x) > end(y)) {
              rv2 = x
              start(rv2) = end(y)
              end(rv2) = end(x)
              rv <- c(rv, rv2)         
              }
          }    
     if(start(x) > start(y)) { 
              if (end(x) > end(y)) {
              rv = x
              start(rv) = end(y)
              end(rv) = end(x)
              }
             }
# if return rv is 0-1, then no interval left from x to be remian            
     return(rv)
}

getridBlackInterval <- function(mypromoter, BK)
{
 o = findOverlaps(mypromoter, BK, type="within")
 if(length(o)>0) {
  grl1 = split(mypromoter[queryHits(o)], 1:length(o)) # You can't mendoapply on a GRanges object
  grl2 = split(BK[subjectHits(o)], 1:length(o))
  mypromoter <- mypromoter[-queryHits(o)]
  
  keptInterval <- unlist(mendoapply(RemoveInterval, grl1, grl2))
  if(any(width(keptInterval) == 2)) keptInterval <- keptInterval[-which(width(keptInterval)==2)]

  if(length(keptInterval>0)) {
   mypromoter <- c(mypromoter, keptInterval)
    }  
  }
 return(mypromoter)
 }




findOverlaps(mypromoter, BK)    # 16 hits
mypromoter <- getridBlackInterval(mypromoter, BK)
hits <- findOverlaps(mypromoter, ignore.strand=TRUE)
(x <- which(duplicated(queryHits(hits))))  # 160, 162
mypromoter <- mypromoter[-x,]



myfile <- paste0('outPut.Rdata/GSE115575/',length(mypromoter),'_GSE115575_500_200_noBlacklist.bed')
df <- data.frame(seqnames=seqnames(mypromoter),
      starts=start(mypromoter)-1,
      ends=end(mypromoter),
      names=names(mypromoter),
   #   scores=elementMetadata(mygr)$Conc,
      strands=strand(mypromoter))
write.table(df, file=myfile, quote=F, sep="\t", row.names=F, col.names=F)


#Homer: findMotifs.pl  237_promoter_500_200_noBlacklist.bed hg38 GSE115575_500_200/ 

#Homer: scanMotifGenomeWide.pl CarlosDesktop/homerv4.10.4/Promoter_500_200/homerResults/motif2.motif hg38 -bed > GLIS3_hg38

### find the instance for Glis motif among the 237 CTS promotes
library(rtracklayer)
Glis <- import('F:/projects/BioTIP/result/Antonio_Feliciano/GSE115575_Holly/GLIS3_hg38.bed')
x <- findOverlaps(mypromoter, Glis)
Glis_Cardiac.Progenitor <- mypromoter[queryHits(x),]
export(Glis_Cardiac.Progenitor,"F:/projects/BioTIP/result/Antonio_Feliciano/GSE115575_Holly/Glis_CP.promoter500.200.bed", format='bed')



