# This code performs CTS analysis on the time-course gene expressional data of GSE115575, using BioTIP.
# Generated by: Antonio Feliciano  <antgabfel_at_gmail.com>
# Verified by Xinan H Yang
# Last update: 03/14/2020

setwd("F:/projects/BioTIP/doc/2020_/Applications/")

# =====================================================================================
# Install and load packages (for new R user) ==========================================
# ======================================================================================
# The following initializes usage of Bioc devel
BiocManager::install(version='devel')
BiocManager::install("BioTIP")
library(BioTIP)
packageVersion("BioTIP")  #[1] ‘1.1.0’


if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

BiocManager::install("edgeR")
BiocManager::install("GenomeInfoDb")
BiocManager::install("GenomicRanges")
BiocManager::install("IRanges")
BiocManager::install("clusterProfiler")
BiocManager::install("gplots")
BiocManager::install("org.Hs.eg.db")
BiocManager::install("org.Mm.eg.db")
BiocManager::install("seq2pathway")
BiocManager::install("GO.db")
install.packages("DBI")

library('edgeR')
library("GenomeInfoDb")
library("GenomicRanges")
library("IRanges")
library('clusterProfiler')
library('gplots')
library('org.Hs.eg.db')
#library('seq2pathway')
#library('GO.db')
library('DBI')

# F:\projects\BioTIP\result\Zhezhen_desktop\code
#source("F:/projects/BioTIP/source/BioTIP_update_3.3_02282020.R")  # combined functions with parameters;
  

# ==============================================================================================
# Load data,  Dataset-specific , Do nOT repeat =============================================
# ==============================================================================================

# GEO_file <- read.delim('input.Rdata/GSE115575/GSE115575_human_total_RNAseq.count.table.txt',
#             header = TRUE, sep = '\t')
# rownames(GEO_file) <- c(1:nrow(GEO_file))
# dim(GEO_file)
# # [1] 61564    27
# 
# raw_counts <- data.frame(GEO_file[, 7:27])
# dim(raw_counts)
# # [1] 61564    21

series_matrix <- read.delim('input.Rdata/GSE115575/Processed_Series_Matrix.txt', header = TRUE, 
                             sep = '\t',row.names = 1)
dim(series_matrix)
# [1] 21  4
 
#all(rownames(series_matrix) %in% colnames(raw_counts)) #TRUE
 
# annotation <- data.frame(GEO_file[, 1:6])
# dim(annotation)
# # [1] 61564     6
# rm(GEO_file) 



# ==============================================================================================
# Filter noisea,  Dataset-specific , Do nOT repeat =============================================
# ==============================================================================================
# 
# # Filter out transcripts whose counts are 0 across all samples ---------------------------------
# table(is.na(cpm(raw_counts, log = TRUE)))
# #   FALSE 
# # 1292844
# x <- apply(raw_counts, 1, sum)
# table(x == 0)
# # FALSE  TRUE 
# # 60073  1491
# raw_counts <- subset(raw_counts, x > 0)
# dim(raw_counts)
# # [1] 60073    21
# 
# # Select sample IDs of each state --------------------------------------------------------------
# ESC_D0 <- rownames(series_matrix)[which(series_matrix$State == 'embryonic stem cells')]
# EMD_D1 <- rownames(series_matrix)[which(series_matrix$State == 'early mesoderm cells')]
# LMD_D2 <- rownames(series_matrix)[which(series_matrix$State == 'late mesoderm cells')]
# CMD_D4 <- rownames(series_matrix)[which(series_matrix$State == 'cardiac mesoderm cells')]
# CPG_D6 <- rownames(series_matrix)[which(series_matrix$State == 'cardiac progenitor cells')]
# ECM_D8 <- rownames(series_matrix)[which(series_matrix$State == 'early cardiomyocytes')]
# LCM_D14 <- rownames(series_matrix)[which(series_matrix$State == 'late cardiomyocytes')]
# 
# # Filter lowly expressed transcripts -----------------------------------------------------------
# cutoff <- -4 # Remaining transcripts should have log2(CPM) > -4
# log2_CPM <- cpm(raw_counts, log = TRUE)
#  # filtering based off of log2(CPM), but after filtering out non-expressed transcripts.
# 
# # Filter across embryonic stem cell samples
# x <- apply(log2_CPM[, ESC_D0], 1, mean)
# table(x > cutoff)
# # FALSE  TRUE 
# # 21254 38819
# x[1:4]
# #           1           2           3           4 
# # -1.27498868  5.24330000 -0.09202896  0.01307455
# y <- names(x)[x > cutoff]
# length(y)
# # [1] 38819
# y[1:4]
# # [1] "1" "2" "3" "4"
# 
# # Filter across early mesoderm cell samples
# x <- apply(log2_CPM[, EMD_D1], 1, mean)
# table(x > cutoff)
# # FALSE  TRUE 
# # 20642 39431
# y <- unique(c(y, names(x)[x > cutoff]))
# length(y)
# # [1] 41052
# 
# # Filter across late mesoderm cell samples
# x <- apply(log2_CPM[, LMD_D2], 1, mean)
# table(x > cutoff)
# # FALSE  TRUE 
# # 20044 40029
# y <- unique(c(y, names(x)[x > cutoff]))
# length(y)
# # [1] 42690
# 
# # Filter across cardiac mesoderm cell samples
# x <- apply(log2_CPM[, CMD_D4], 1, mean)
# table(x > cutoff)
# # FALSE  TRUE 
# # 19837 40236
# y <- unique(c(y, names(x)[x > cutoff]))
# length(y)
# # [1] 44362
# 
# # Filter across cardiac progenitor cell samples
# x <- apply(log2_CPM[, CPG_D6], 1, mean)
# table(x > cutoff)
# # FALSE  TRUE 
# # 19857 40216
# y <- unique(c(y, names(x)[x > cutoff]))
# length(y)
# # [1] 45181
# 
# # Filter across early cardiomyocyte samples
# x <- apply(log2_CPM[, ECM_D8], 1, mean)
# table(x > cutoff)
# # FALSE  TRUE 
# # 20583 39490
# y <- unique(c(y, names(x)[x > cutoff]))
# length(y)
# # [1] 45490
# 
# # Filter across late cardiomyocyte samples
# x <- apply(log2_CPM[, LCM_D14], 1, mean)
# table(x > cutoff)
# # FALSE  TRUE 
# # 21238 38835
# y <- unique(c(y, names(x)[x > cutoff]))
# length(y)
# # [1] 45898
# 
# # Apply filter across all samples
# filtered_log2_CPM <- log2_CPM[y, ]
# dim(filtered_log2_CPM)
# # [1] 45898    21
# 
# # Results --------------------------------------------------------------------------------------
# # Total of 46k transcripts selected from all 7 states
# 
# # Check how many cardiac lncRNAs are remaining
# cardiac_lncRNAs <- c('MEG3', 'GATA6-AS', 'XIST', 'FENDRR')
# for (each in cardiac_lncRNAs) {
#   print(each)
#   print((tmp <- rownames(annotation)[grep(each, annotation$name)]))
#   print(tmp %in% rownames(filtered_log2_CPM))
#   print(log2_CPM[tmp, ])
# }
# MEG3: 3/3, GATA6-AS: 1/1, XIST: 1/2, FENDRR: 1/1


# ============================================================================
# Normalize data, Dataset-specific , DO NOT repeat   =========================
# ============================================================================
#load('input.Rdata/GENCODEv25_gtf_summary.RData')
#dim(hg38)
# [1] 2579816      10

# # TMM ------------------------------------------------------------------------------------------
# filtered_raw_counts <- raw_counts[y, ]
# dim(filtered_raw_counts)
# # [1] 45898    21
# TMM <- calcNormFactors(DGEList(filtered_raw_counts), method = c("TMM"))
# TMM_log2_CPM <- cpm(TMM, log = TRUE)
# 
# min(TMM_log2_CPM) # -5.472802
# max(TMM_log2_CPM) # 17.49137
# 
# 
# # Justification for additional normalization with cyclic loess
# expressed_transcripts <- as.data.frame(TMM_log2_CPM >= -5)
# table(TMM_log2_CPM >= -5)
# # FALSE   TRUE
# # 42293 921565
# N <- rowSums(expressed_transcripts)
# expressed_transcripts$N <- N
# hist(N, 100)
# lowly_expressed_id <- rownames(expressed_transcripts)[which(expressed_transcripts$N <= 15)]
# lowly_expressed_name <- unique(annotation[lowly_expressed_id, "name"])
# length(lowly_expressed_name) # 2914
# 
# hly_expressed_id <- rownames(expressed_transcripts)[which(expressed_transcripts$N > 15)]
# hly_expressed_name <- unique(annotation[hly_expressed_id, "name"])
# length(hly_expressed_name) # 41928
# 
# head(hg38)
# table(hg38[, 8])
# coding <- subset(hg38, hg38[, 8] == 'protein_coding')
# 1- nrow(coding)/nrow(hg38)  #[1] 0.087 coding on transcript levels (NOT used)
# symbol <- unique(coding[, 7])
# length(symbol) # 19903
# 1 - 19903/length(unique(hg38[,7])) # [1] 0.646 coding on symbol levels
# 
# # Count based on the symbols, the % of symbols are highly/lowly expressed
# tmp <-  table(symbol %in% unique(as.vector(lowly_expressed_name)))
# tmp[2]/sum(tmp)  # 1.7% of coding gene symbols
# tmp <-  table(symbol %in% unique(as.vector(hly_expressed_name)))
# tmp[2]/sum(tmp)  # 91.6% of coding gene symbols
# # background
# tmp <- table(symbol %in% unique(c(as.vector(hly_expressed_name),
#                                   as.vector(lowly_expressed_name))))
# tmp[2]/sum(tmp)  # 93.2%
# 
# # Counts on the transcripts (i.e., % of hig_transcripts are (coding) symbols) 
# tmp3 <- table(lowly_expressed_name %in% symbol) # 11.4% coding, 88.6% non-coding
# tmp4 <- table(hly_expressed_name %in% symbol) # 43.5% coding, 56.5% non-coding
# tmp <- table(unique(c(as.vector(hly_expressed_name),as.vector(lowly_expressed_name)))
#              %in% symbol) # 41.4%% coding, bk
# tmp[2]/sum(tmp)
# 
# rm(tmp, tmp3, tmp4)
# gc()

# Cyclic loess ---------------------------------------------------------------------------------
# cyclic_loess <- normalizeCyclicLoess(TMM_log2_CPM, iterations = 10)
# save(cyclic_loess, file = 'outPut.Rdata/GSE115575/cyclic_loess.RData')

#load(file = 'input.Rdata/GSE115575/cyclic_loess.RData')

#limma::plotMA(cyclic_loess, main = "Normalized Counts")
#dev.copy2pdf(file = 'ROutputFigs/GSE115575/MAplot_cyclic_loess.pdf')

# par(mfrow=c(4,2))
# hist(log2_CPM, 100, main = "log2 CPM of 60,073 transcripts")
# abline(v = cutoff, lty = 2, col='red')
# boxplot(log2_CPM, main = "log2 CPM of 60,073 transcripts", xlab = 'samples')
# hist(filtered_log2_CPM, 100, main = "log2 CPM of 45,898 transcripts")
# boxplot(filtered_log2_CPM, main = "log2 CPM of 45,898 transcripts", xlab = 'samples')
# hist(TMM_log2_CPM, 100, main = "TMM")
# boxplot(TMM_log2_CPM, main = "TMM", ylab = 'log2 CPM', xlab = 'samples')
# hist(cyclic_loess, 100, main = "Normalized Counts", xlab = 'Normalized expression levels')
# boxplot(cyclic_loess, main = paste("Normalized Counts, cut = ", cutoff, ", ", length(y),
#         "transcripts"), xlab = 'Samples', ylab = 'Normalized expression levels')
# 
# dev.copy2pdf(file = 'ROutputFigs/GSE115575/hist_box_filter_TMM_cyclic.pdf')

#rm(log2_CPM, filtered_log2_CPM, TMM_log2_CPM )

# ==============================================================================================
# Analyze data =================================================================================
# ==============================================================================================

load('input.Rdata/GSE115575/cyclic_loess.RData')

# NPS on cyclic loess-normalized data ----------------------------------------------------------
samplesL <- split(rownames(series_matrix), f = factor(series_matrix$State, 
                  levels = c('embryonic stem cells', 'early mesoderm cells',
                  'late mesoderm cells', 'cardiac mesoderm cells', 'cardiac progenitor cells',
                  'early cardiomyocytes', 'late cardiomyocytes')))
unlist(lapply(samplesL, length))
# embryonic stem cells     early mesoderm cells      late mesoderm cells 
#       3                        3                        3 
# cardiac mesoderm cells cardiac progenitor cells     early cardiomyocytes 
#       3                        3                        3 
# late cardiomyocytes 
#       3 

test <- sd_selection(cyclic_loess, samplesL, 0.01) # Standard deviations per state
par(mfrow = c(1, 1))
boxplot(sapply(test, function(x) apply(x, 1, sd)), main = 'SD Selection',
        ylab = 'Standard deviation')
dev.copy2pdf(file = 'ROutputFigs/GSE115575/standard_deviations.pdf')

igraphL <- getNetwork(test, fdr = 1) # Networks per state
# [1] "cardiac mesoderm cells:458 nodes"
# [1] "cardiac progenitor cells:458 nodes"
# [1] "early cardiomyocytes:458 nodes"
# [1] "early mesoderm cells:458 nodes"
# [1] "embryonic stem cells:458 nodes"
# [1] "late cardiomyocytes:458 nodes"
# [1] "late mesoderm cells:458 nodes"
cluster <- getCluster_methods(igraphL)
#par(mfrow = c(2, 4))
#for (i in names(igraphL)) {
#  plot(igraphL[[i]], mark.groups = cluster[[i]], vertex.label = NA, vertex.size = 3,
#       edges.color = 'grey90')
#}
#dev.copy2pdf(file = 'ROutputFigs/GSE115575/networks.pdf')

# MCI-score
membersL_weight <- getMCI(cluster, test, adjust.size = TRUE) # Weighted
plotBar_MCI(membersL_weight, ylim = c(0, 40))
dev.copy2pdf(file = 'ROutputFigs/GSE115575/weighted_MCI.pdf')

maxMCIms <- getMaxMCImember(membersL_weight[[1]], membersL_weight[[2]], min = 10) # Weighted
maxMCI <- getMaxStats(membersL_weight[[2]], maxMCIms[[1]])
CTS <- getCTS(maxMCI, maxMCIms$members)
# [1] "Length: 253"

# Save identified signatures for downstream motif analysis
weighted_cyclic_loess_signature <- annotation[CTS, ]
save(weighted_cyclic_loess_signature, file = 'outPut.Rdata/GSE115575/weighted_cyclic_loess_signature.RData')

head(weighted_cyclic_loess_signature)
#     chrom    start      end         name score strand
#6     chr1    34554    36081      FAM138A     .      -
#935   chr1 28031884 28032180    RN7SL559P     .      -
#950   chr1 28453541 28453934  RP1-308E4.1     .      +
#1477  chr1 46134531 46139081  RP4-533D7.5     .      +
#1616  chr1 52150104 52150419    RN7SL788P     .      +
#1871  chr1 65066627 65067737 RP4-535B20.1     .      -


# MCI simulation 
## NOTE that this step taks a while !!!!
load(file = 'outPut.Rdata/GSE115575/weighted_cyclic_loess_signature.RData')
CTS <- rownames(weighted_cyclic_loess_signature)

simMCI <- simulationMCI(length(CTS), samplesL, cyclic_loess,
                        adjust.size = TRUE)
gc()
#save(simMCI, file="F:/projects/BioTIP/result/Antonio_Feliciano/GSE115575_Holly/CTS_simu.RData", compress=TRUE )  #!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

#load(file="F:/projects/BioTIP/result/Antonio_Feliciano/GSE115575_Holly/CTS_simu.RData")
plot_MCI_Simulation(maxMCI, simMCI)
dev.copy2pdf(file = 'ROutputFigs/GSE115575/GenePermutation_weighted_MCI.pdf')


 
## estimate significance (classical Ic-score is used because the sample sizes are equal) ########################################
######  Ic score, shulffing genes ##############
C= 1000
n <- length(CTS)
Ic_score <- getIc(cyclic_loess, samplesL, CTS, fun="cor")
Ic_score
# embryonic stem cells     early mesoderm cells      late mesoderm cells 
#   0.6851250                0.6643228                0.6777760 
# cardiac mesoderm cells cardiac progenitor cells     early cardiomyocytes 
#   0.6843066                1.2247229                0.6968536 
# late cardiomyocytes 
#   0.6772960 
simuIc_g  <- simulation_Ic(n, samplesL,cyclic_loess, B=C,
                           fun='cor')


#### Ic score,  shulffing labelling
x <- 'cardiac progenitor cells'
simuIc_s <- matrix(nrow=length(samplesL), ncol=C)
rownames(simuIc_s) = names(samplesL)
for(j in 1:length(samplesL)) {
  ns <- length(samplesL[[j]])  # of cells
  simuIc_s[j,] <- simulation_Ic_sample(cyclic_loess, ns, Ic=Ic_score[x],
                                       genes=CTS, B=C,
                                       fun="cor")
}

# save( simuIc_g,  simuIc_s, file="F:/projects/BioTIP/result/Antonio_Feliciano/GSE115575_Holly/simuIC.RData", compress=TRUE)

# load(file="F:/projects/BioTIP/result/Antonio_Feliciano/GSE115575_Holly/simuIC.RData")  

par(mfrow=c(2,2))

plot_Ic_Simulation(Ic_score, simuIc_g, las = 2, ylab="Ic",
                   main=paste(length(CTS),x, "genes", "\n","vs. ",C, "gene-permutations"),
                   fun="boxplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_g, 
                   main = paste("Delta IC",n,"random genes"), 
                   ylab=NULL)
plot_Ic_Simulation(Ic_score, simuIc_s, las = 2, ylab="Ic",ylim=c(0.6,1.3),
                   main=paste(length(CTS), x, "genes", "\n","vs.", C, "sample-permutations"),
                   fun="matplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_s, 
                   main = "Delta IC, labelling shuffling", 
                   ylab=NULL)

dev.copy2pdf(file=paste0("ROutputFigs/GSE115575/Ic_vsSimulation_",length(CTS),"CTS.pdf"))




######  new Ic score, shulffing genes ##############
C= 1000
n <- length(CTS)
Ic_score <- getIc(cyclic_loess, samplesL, CTS, fun="BioTIP")
Ic_score
# embryonic stem cells     early mesoderm cells      late mesoderm cells 
#   0.3574008                0.3377966                0.3485601 
# cardiac mesoderm cells cardiac progenitor cells     early cardiomyocytes 
#   0.3536546                1.0514114                0.3639117 
# late cardiomyocytes 
#   0.3497598 
simuIc_g  <- simulation_Ic(n, samplesL,cyclic_loess, B=C,
                           fun='BioTIP')


#### Ic score,  shulffing labelling
x <- 'cardiac progenitor cells'
simuIc_s <- matrix(nrow=length(samplesL), ncol=C)
rownames(simuIc_s) = names(samplesL)
for(j in 1:length(samplesL)) {
  ns <- length(samplesL[[j]])  # of cells
  simuIc_s[j,] <- simulation_Ic_sample(cyclic_loess, ns, Ic=Ic_score[x],
                                       genes=CTS, B=C,
                                       fun="BioTIP")
}

# save( simuIc_g,  simuIc_s, file="F:/projects/BioTIP/result/Antonio_Feliciano/GSE115575_Holly/simuBioTIP.RData", compress=TRUE)

# load(file="F:/projects/BioTIP/result/Antonio_Feliciano/GSE115575_Holly/simuBioTIP.RData")  

par(mfrow=c(2,2))

plot_Ic_Simulation(Ic_score, simuIc_g, las = 2, ylab="IC*", ylim=c(0.3,0.6),
                   main=paste(length(CTS),x, "genes", "\n","vs. ",C, "gene-permutations"),
                   fun="boxplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_g, 
                   main = paste("Larget - 2nd largest",n,"genes"), 
                   ylab=NULL)
plot_Ic_Simulation(Ic_score, simuIc_s, las = 2, ylab="Ic*",ylim=c(0.6,1.3),
                   main=paste(length(CTS), x,"genes", "\n","vs. ",C, "sample-permutations"),
                   fun="matplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_s, 
                   main = "Delta IC*, labelling shuffling", 
                   ylab=NULL)

dev.copy2pdf(file=paste0("ROutputFigs/GSE115575/BioTIP_vsSimulation_",length(CTS),"CTS.pdf"))







 ###### motif analysis #####
 source('Rcode/GSE115575_motif_analysis_CTS_acardiacProgenitor_xy.R')








































































































