# This code performs CTS analysis on the time-course gene expressional data of GSE2565, using BioTIP.
# Generated by: Zhezhen Wang
# Revised and annotated by: Xinan H Yang (xyang2_at_uchicago.edu)
# Last update: 03/06/2020

# The following initializes usage of Bioc devel
BiocManager::install(version='devel')
BiocManager::install("BioTIP")
library(BioTIP)
packageVersion("BioTIP")  #[1] 1.1.1

library(stringr)
library(psych)
library(igraph)
 
# instead of the following one command
#source("F:/projects/BioTIP/source/BioTIP_update_3.3_02282020.R")  # combined functions with parameters;
setwd("F:/projects/BioTIP/doc/2020_/Applications/")


GSE2565 = read.table("input.Rdata/GSE2565/GSE2565_series_matrix.txt",head = T,sep = '\t',comment = '!')
row.names(GSE2565) = GSE2565[,1]
GSE2565 = as.matrix(GSE2565[,-1])
dim(GSE2565 ) #[1] 22690   104

max(GSE2565)  #[1] 11586.8  not log-transformed yet
df = log2(GSE2565+1)
summary(df)

## build the data structure (nested list) for clinical state information
cli = t(read.delim("input.Rdata/GSE2565/cli.txt"))
colnames(cli) = cli[1,]
cli = cli[-1,]
colnames(cli)[12] = 'expose'
colnames(cli)[14] = 'time_point'
table(cli[,"expose"])
#  Air   CG None 
#  50   48    6 

####################### BioTIP application  #####################

cliL = list('control' = subset(cli,cli[,12] != 'CG'), 'treated' = subset(cli, cli[,"expose"] != 'Air'))
intersect(cliL[['treated']][,'!Sample_geo_accession'], cliL[['control']][,'!Sample_geo_accession'])
#[1] "GSM49374" "GSM49375" "GSM49376" "GSM49377" "GSM49378" "GSM49379"       # 6 tome 0 sampels are re-used
sampleL = lapply(cliL, function(x) split(x[,'!Sample_geo_accession'], f = x[,'time_point']))
unlist(lapply(sampleL, length))
#control treated 
#      9       9 
 unlist(lapply(sampleL[['control']], length))
#  0 0.5   1  12  24   4  48  72   8 
#  6   6   6   6   6   6   8   6   6 
 unlist(lapply(sampleL[['treated']], length))
#  0 0.5   1  12  24   4  48  72   8 
#  6   6   6   6   6   6   6   6   6 

order = c('0','0.5','1','4','8','12','24','48','72')
sampleL[['control']] = sampleL[['control']][order]
sampleL[['treated']] = sampleL[['treated']][order]

 intersect(unlist(sampleL[['treated']]), unlist(sampleL[['control']]))
#[1] "GSM49374" "GSM49375" "GSM49376" "GSM49377" "GSM49378" "GSM49379"

## step 1) pre-select the transcrips
## optimation is ignored in this trial due to small sample size
## 'longitudinal reference' strategy is applied taking advantage of the control time-series

 l = list()
 sl=list()
 j=0.2; k=50  
 c_df = df[,do.call(c, sampleL[['control']])]
 tr_df = df[,do.call(c, sampleL[['treated']])]

## Preselect the top 2% transcripts with the highest relative standard deviation
## in each time point compared to transcript in the correspong control time point
 test = sd_selection(df=tr_df, samplesL=sampleL[['treated']], cutoff= 0.02, 
                     method = 'longitudinal reference', 
                     control_df = c_df, 
                     control_samplesL = sampleL[['control']])

## step 2.1) Build gene regulatory network based on expresional PCC
 igraphL = getNetwork(test, fdr=j)
#[1] "0:0 nodes"
#[1] "0.5:178 nodes"
#[1] "1:361 nodes"
#[1] "4:438 nodes"
#[1] "8:446 nodes"
#[1] "12:439 nodes"
#[1] "24:446 nodes"
#[1] "48:369 nodes"
#[1] "72:105 nodes"

## step 2.2) Network partition using random walk
## this step constructs modules (geoups of transcriopts) based on their expression levels
cluster = getCluster(igraphL)

## step 3.1) calculate MCI score for each module
membersL_treated = getMCI(cluster,test, adjust.size=TRUE)
names(membersL_treated)
#[1] "members" "MCI"     "sd"      "PCC"     "PCCo"
plotBar_MCI(membersL_treated, min = k, ylim=c(0,20))
dev.copy2pdf(file = 'ROutputFigs/GSE2565/MCI_fdr0.2min50.pdf')

## step 3.2) identify biomodule per state
maxMCIms = getMaxMCImember(membersL_treated[["members"]],membersL_treated[["MCI"]],min= k)
maxMCI = getMaxStats(membersL_treated[["MCI"]],maxMCIms[["idx"]])
maxMCI
#         1         4         8        12        24        48 
# 10.519391 10.340600 19.124593 13.128889 12.444359  8.812538

## get biomodule with the highest MCI at time-point 8
 tmp <- table(membersL_treated[['members']][['8']])
 tmp[tmp>k]
#   2   9  
# 172  56  

CTS = names(membersL_treated[['members']][['8']])[membersL_treated[['members']][['8']]==2]
length(CTS) # 172
save(CTS, file= 'outPut.Rdata/GSE2565/CTS_172probeID.GSE2565_fdr0.2min50.rData')


## step 3.3) check the significance of CTS by MCI scoring system 

simMCI_treated = simulationMCI(length(CTS),sampleL[["treated"]], tr_df, adjust.size =TRUE, B=100)
simMCI_control = simulationMCI(length(CTS),sampleL[["control"]], c_df, adjust.size =TRUE, B=100)




## step 4.1) Caslculate the ic-score and evaluate its significance

Ic_treated = getIc(tr_df,sampleL[["treated"]],CTS)
simIc_treated = simulation_Ic(length(CTS),sampleL[["treated"]], tr_df, B=1000)

Ic_control = getIc(c_df,sampleL[["control"]],CTS)
simIc_control = simulation_Ic(length(CTS),sampleL[["control"]], c_df, B=1000)

## plot 
load(file='outPut.Rdata/GSE2565/MCI_Ic_genepermutation_GSE256_BioTIP171g.RData')

quickMCI = function(df,sampleL,probeset){
  countsL= lapply(sampleL[['control']], function(x) df[,x])
  names(countsL) = names(sampleL[['control']])
  x='8'
  m = subset(countsL[[x]],row.names(countsL[[x]]) %in% probeset)
  comple = subset(countsL[[x]],!row.names(countsL[[x]]) %in% probeset)
  
  #PCCo = mapply(function(x,y) abs(cor(t(x),t(y))),comple,m)
  PCCo = abs(cor(t(comple),t(m)))
  PCCo_avg = mean(PCCo) 
  PCC = abs(cor(t(m)))
  
  PCC_avg = (sum(PCC,na.rm = T)-nrow(PCC))/(nrow(PCC)^2-nrow(PCC))
  
  sd_chen = apply(m,1,sd)
  CI_chen = mean(sd_chen)*(PCC_avg/PCCo_avg)*sqrt(nrow(m))
  return(CI_chen)
}

MCI_treated <-  maxMCI[['8']]     # [1] 19.12459
MCI_control <-  quickMCI(df,sampleL,CTS)   # [1] 2.037948
names(MCI_treated) <- names(MCI_control) <- '8'

par(mfrow = c(2,2))
plot_MCI_Simulation(MCI_treated,simMCI_treated,
                    order = c('0','0.5','1','4','8','12','24','48','72'),
                    main= paste("Treated", length(CTS), "100 gene permutations"))
#lines(maxMCI, col='red')
plot_MCI_Simulation(MCI_control,simMCI_control,
                    order = c('0','0.5','1','4','8','12','24','48','72'),
                    main= paste("Control", length(CTS), "100 gene permutations"))

plot_Ic_Simulation(Ic_treated,simIc_treated,
                   order = c('0','0.5','1','4','8','12','24','48','72'),
                   main= paste("Treated", length(CTS), "1000 gene permutations"))
plot_Ic_Simulation(Ic_control,simIc_control,
                   order = c('0','0.5','1','4','8','12','24','48','72'),
                   main= paste("Control", length(CTS), "1000 gene permutations"))

dev.copy2pdf(file = 'ROutputFigs/GSE2565/MCI_Ic_genepermutation_GSE256_BioTIP171g.pdf')

#### step 4.2) Ic score,  shulffing labelling
x <- '8'
C=1000
#ns <- length(sampleL[["treated"]][[x]])  # of cells
simuIc_treated_s <- matrix(nrow=length(sampleL[["treated"]]), ncol=C)
rownames(simuIc_treated_s) = names(sampleL[["treated"]])
for(j in 1:length(sampleL[["treated"]])) {
  ns <- length(sampleL[["treated"]][[j]])
  simuIc_treated_s[j,] <- simulation_Ic_sample(tr_df, ns, Ic=Ic_treated,
                                       genes=CTS,  B=C,
                                       fun="cor")
}
# ns.c <- length(sampleL[["control"]][[x]])  # of cells
simuIc_control_s <- matrix(nrow=length(sampleL[["control"]]), ncol=C)
rownames(simuIc_control_s) = names(sampleL[["control"]])
for(j in 1:length(sampleL[["control"]])) {
  ns.c <- length(sampleL[["control"]][[j]])
  simuIc_control_s[j,] <- simulation_Ic_sample(c_df, ns.c, Ic=Ic_control,
                                               genes=CTS,  B=C,
                                               fun="cor")
}


save(simMCI_treated,simMCI_control, Ic_treated, Ic_control,
     simIc_treated, simIc_control, 
     simuIc_treated_s, simuIc_control_s,
     file='outPut.Rdata/GSE2565/MCI_Ic_genepermutation_GSE256_BioTIP171g.RData', compress=TRUE)


## plot 
n <- length(CTS)
par(mfrow=c(2,2))
Ic_score <- Ic_treated
simuIc_g <- simIc_treated
simuIc_s <- simuIc_treated_s
plot_Ic_Simulation(Ic_score, simuIc_g, las = 2, ylab="Ic",
                   main=paste(length(CTS),x, "h-treated genes", "\n","vs. ",C, "gene-permutations"),
                   fun="boxplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_g, 
                   main = paste("Larget - 2nd largest",n,"genes"))
plot_Ic_Simulation(Ic_score, simuIc_s, las = 2, ylab="Ic",ylim=c(0.4,1.1),
                   main=paste(length(CTS), x,"h-treated genes", "\n","vs. ",C, "sample-permutations"),
                   fun="matplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_s, 
                   main = paste("Delta Ic, sample shuffling","\n", "treated samples"))

dev.copy2pdf(file=paste0("ROutputFigs/GSE2565/treated_Ic_vsSimulation_",length(CTS),"CTS.pdf"))


par(mfrow=c(2,2))
Ic_score <- Ic_control
simuIc_g <- simIc_control
simuIc_s <- simuIc_control_s
plot_Ic_Simulation(Ic_score, simuIc_g, las = 2, ylab="Ic", ylim=c(0.3, 0.7), 
                   main=paste(length(CTS),x, "h-treated genes", "\n","vs. ",C, "gene-permutations"),
                   fun="boxplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_g, 
                   main = paste("Larget - 2nd largest",n,"genes"))
plot_Ic_Simulation(Ic_score, simuIc_s, las = 2, ylab="Ic",ylim=c(0.3,0.7),
                   main=paste(length(CTS), x,"h-treated genes", "\n","vs. ",C, "sample-permutations"),
                   fun="matplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_s, 
                   main = paste("Delta Ic, sample shuffling","\n", "control samples"))

dev.copy2pdf(file=paste0("ROutputFigs/GSE2565/control_Ic_vsSimulation_",length(CTS),"CTS.pdf"))




## step 4.3) further check the significance of CTS by the new Ic scores
## This step is not necessary as the sammple size are equalamong states

Ic_adjusted = getIc(tr_df,sampleL[["treated"]],CTS, fun ='BioTIP')
simIc_adjusted = simulation_Ic(length(CTS),sampleL[["treated"]], tr_df,  fun ='BioTIP')

Ic_adjusted_control = getIc(c_df,sampleL[["control"]],CTS, fun ='BioTIP')
simIc_adjusted_control = simulation_Ic(length(CTS),sampleL[["control"]], c_df,  fun ='BioTIP')

# save(Ic_adjusted, simIc_adjusted, Ic_adjusted_control, simIc_adjusted_control,
# file="F:/projects/BioTIP/result/GSE2565/newIc_genepermutation_GSE256_BioTIP171g.RData")


par(mfrow = c(1,2))
plot_Ic_Simulation(Ic_adjusted, simIc_adjusted,  las=2, ylab='new Ic*', ylim=c(0,0.8),
   order = c('0','0.5','1','4','8','12','24','48','72'),
   main= paste("Treated:", length(CTS), "1000 gene permutations") )
plot_Ic_Simulation(Ic_adjusted_control, simIc_adjusted_control,  las=2, ylab='new Ic*', ylim=c(0,0.8),
   order = c('0','0.5','1','4','8','12','24','48','72'),
   main= paste("Control:", length(CTS), "1000 gene permutations") )
   
dev.copy2pdf(file = 'ROutputFigs/GSE2565/BioTIP_GSE256_BioTIP171g.pdf')

par(mfrow = c(1,2))
plot_Ic_Simulation(Ic_adjusted, simIc_adjusted,  las=2, ylab='new Ic*', ylim=c(0,0.8),
                   order = c('0','0.5','1','4','8','12','24','48','72'),
                   main= paste("Treated:", length(CTS), "1000 gene permutations"),fun="boxplot" )
plot_Ic_Simulation(Ic_adjusted_control, simIc_adjusted_control,  las=2, ylab='new Ic*', ylim=c(0,0.8),
                   order = c('0','0.5','1','4','8','12','24','48','72'),
                   main= paste("Control:", length(CTS), "1000 gene permutations"),fun="boxplot" )

dev.copy2pdf(file = 'ROutputFigs/GSE2565/BioTIP_GSE256_BioTIP171g_boxplot.pdf')





####################################################
## Compare to the 259 genes reported by Chen 2012  ##
####################################################
## DNB_lung_injur_Symbols.txt published 262 gene symbols (GSE2565_chen_Ic_CI.R)
## manually added in HP,corrected ENSMUSG00000040078, changed C-MYC to MYC,E2F1-1 to E2F1,JUND1 to JUND
dnb_chen = read.table("input.Rdata/GSE2565/dnb_clean.GSE2565.txt",head = T)
dnb_chen <- unique(as.vector(dnb_chen[,1]))
length(dnb_chen)  #[1] 259  gene symbols

probesetid = read.delim("input.Rdata/GSE2565/GPL8321-17396.txt",comment = '#',head = T)
table(dnb_chen %in% toupper(probesetid$Gene.Symbol))
#FALSE  TRUE 
#   31   228 
tmp = dnb_chen[!dnb_chen %in% toupper(probesetid$Gene.Symbol)]
 
reannot = read.table("input.Rdata/GSE2565/reannotated_GSE3565.txt")
dnb_symbol = c(as.character(dnb_chen),as.character(reannot[,1]))

table(dnb_symbol %in% toupper(probesetid$Gene.Symbol))
# FALSE  TRUE 
#  31   239
probeset = probesetid[match(dnb_symbol,toupper(probesetid$Gene.Symbol)),'ID']
length(probeset) #[1] 270
table(is.na(probeset))
#FALSE  TRUE 
# 239    31
probeset = probeset[!is.na(probeset)]
length(probeset) #[1] 239

## MCI treated
countsL= lapply(sampleL[["treated"]], function(x) df[,x])
names(countsL) = names(sampleL[["treated"]])
x='8'
m = subset(countsL[[x]],row.names(countsL[[x]]) %in% probeset)
comple = subset(countsL[[x]],!row.names(countsL[[x]]) %in% probeset)
PCCo = abs(cor(t(comple),t(m)))
PCCo_avg = mean(PCCo) 
PCC = abs(cor(t(m)))
PCC_avg = (sum(PCC,na.rm = T)-nrow(PCC))/(nrow(PCC)^2-nrow(PCC)) 
sd_chen = apply(m,1,sd)
CI_chen = mean(sd_chen)*(PCC_avg/PCCo_avg)
#[1] 0.4680342

######## simulation for treated samples 
mCI_chen = simulationCI(length(probeset), sampleL[["treated"]],df, B = 1000) #
row.names(mCI_chen) = names(sampleL[[2]])
save(mCI_chen,file = 'outPut.Rdata/GSE2565/mCI_chen239g.GSE2565_treated.rData')  # 'mCI_chen.GSE2565_DNB.rData'

######  plot treated samples 
names(CI_chen) = '8'
plot_MCI_Simulation(CI_chen, mCI_chen[,1:100], las = 2, 
        order = names(sampleL[["treated"]])[c(1:3,6,9,4,5,7,8)])
dev.copy2pdf(file = 'ROutputFigs/GSE2565/boxplot.GSE2565_MCI_DNBsim.pdf')


## MCI control
countsL= lapply(sampleL[["control"]], function(x) df[,x])
names(countsL) = names(sampleL[["control"]])
x='8'
m = subset(countsL[[x]],row.names(countsL[[x]]) %in% probeset)
comple = subset(countsL[[x]],!row.names(countsL[[x]]) %in% probeset)
PCCo = abs(cor(t(comple),t(m)))
PCCo_avg = mean(PCCo) 
PCC = abs(cor(t(m)))
PCC_avg = (sum(PCC,na.rm = T)-nrow(PCC))/(nrow(PCC)^2-nrow(PCC))
sd_chen = apply(m,1,sd)
CI_chen = mean(sd_chen)*(PCC_avg/PCCo_avg)
names(CI_chen) = '8'

######## MCI simulation for control samples 
mCI_chen = simulationCI(length(probeset), sampleL[["control"]],df,B = 1000) #
row.names(mCI_chen) = names(sampleL[["control"]])
save(mCI_chen,file = 'outPut.Rdata/GSE2565/mCI_chen239g.GSE2565_control.rData')   # 'mCI_chen.GSE2565_DNB_control.rData'

## plot control samples 
plot_MCI_Simulation(CI_chen, mCI_chen[,1:100],las = 0, 
  order = names(sampleL[[1]])[c(1:3,6,9,4,5,7,8)])
#dev.copy2pdf(file = 'lines.GSE2565_CI_contrl_DNBsim.pdf')
dev.copy2pdf(file = 'ROutputFigs/GSE2565/boxplot.GSE2565_MCI_contrl_DNBsim.pdf')


##  Ic for treated  samples
Ic_chen = getIc(tr_df,sampleL[["treated"]],probeset)
Ic_chensim = simulation_Ic(length(probeset),sampleL[["treated"]],tr_df)
plot_Ic_Simulation(Ic_chen, Ic_chensim,
    order = c('0','0.5','1','4','8','12','24','48','72'))
dev.copy2pdf(file = 'ROutputFigs/GSE2565/lines_GSE2565_Ic_chen_sim.pdf')

## Ic for control sampels 
Ic_chen_control = getIc(c_df, sampleL[["control"]], probeset)
Ic_chensim_control = simulation_Ic(length(probeset),sampleL[["control"]],c_df)
plot_Ic_Simulation(Ic_chen_control,Ic_simulation_control,
    order = c('0','0.5','1','4','8','12','24','48','72'),ylim = c(0.3,0.7))
dev.copy2pdf(file = 'ROutputFigs/GSE2565/lines_GSE2565_Ic_chen_sim_contrl.pdf')

