# This code performs CTS analysis on the log2-transformed gene expressional data of GSE6136, using BioTIP.
# Generated by: Zhezhen Wang
# Organized and annotated by: Xinan H Yang (xyang2_at_uchicago.edu)
# Last update: 03/05/2020

# The following initializes usage of Bioc devel
BiocManager::install(version='devel')
BiocManager::install("BioTIP")
library(BioTIP)
packageVersion("BioTIP")  #[1] ‘1.1.0’

library(stringr)
library(psych)
library(igraph)
 
# instead of the following one command
#source("F:/projects/BioTIP/source/BioTIP_update_3.3_02282020.R")  # combined functions with parameters;
setwd("F:/projects/BioTIP/doc/2020_/Applications/")

###################################################################
# original code: F:\ZheZhen\GSE49711\code\DNB\ 
# DNB_GSE49711_relativeMaxV2.R; DNB_4stages_line_dist_plots.R;  DNB_GSE49711_CI_Ic_nonormal.R 
# More details are given in the file CTS.Fig_GSRE49711code.docx 

###################
## load data ######

load('outPut.Rdata/GSE49711/stable.top1relativeSD_morethan80.rData')
stages = c('LR-EFS','ILR-EFS','LR-E','ILR-E','HIR-EFS','HR-EFS','HIR-E','HR-E')
library("Biobase")

cli <- read.table('input.Rdata/GSE49711/full_cli_xy.txt' , header=T)
#tmp = names(table(cli$group_xy))

load("input.Rdata/GSE49711/full_table.simple.rData")
#load("E:\\ZheZhen2019_bk\\GSE49711\\result\\HRLRm\\full_table.MergedTable_methy.rData")
#full_table <- full_table[,1:17]w
expressed = subset(full_table, AveExpr>1)
dim(expressed)  # [1] 48682   219

load("input.Rdata/GSE49711/eSet_log2.rData")
counts = (exprs(eSet))
counts = subset(counts,row.names(counts) %in% expressed$Row.names)
dim(counts) #[1] 48682   498

counts = subset(counts,row.names(counts) %in% subset(full_table,!grepl('\\|',Chromosome))$Row.names)
dim(counts) #[1] 48539   498


##############################
## transcript preselection ###
## This runs a while !!!   ###

samplesL = split(cli$Row.names, f = cli$group_xy)
unlist(lapply(samplesL, length))
#HIR-E HIR-EFS    HR-E  HR-EFS   ILR-E ILR-EFS    LR-E  LR-EFS 
#  11      19     120      56      29      72      23     168 
save(samplesL, file="input.Rdata/GSE49711/GSE49711_samplesL.rData")

stable <- optimize.sd_selection(counts,  samplesL,  B=100,  percent=0.8,  
                                           times=0.8, cutoff=0.01)
lengths(stable)
#LR-EFS ILR-EFS    LR-E   ILR-E HIR-EFS  HR-EFS   HIR-E    HR-E 
#  274     212      72     172     116     155      27     222  
save(stable,file = 'outPut.Rdata/GSE49711/stable.top1relativeSD_morethan80.rData')  
  

stage <- factor(names(stable), levels=c('LR-EFS', 'ILR-EFS', 'HIR-EFS',  'HR-EFS',
                    'LR-E',   'ILR-E',    'HIR-E',    'HR-E')) 
A = lapply(stages, function(x) counts[,as.character(subset(cli, group_xy == x)$Row.names)])
names(A) = stages
A = lapply(stages, function(x) A[[x]][stable[[x]],])
names(A) = stages

sdL = sapply(A,function(x) apply(x,1,sd))

par(mfrow = c(1,2))
EFI = sdL[c(1,2,5,6)]
EI = sdL[c(3,4,7,8)]
boxplot(EFI,outline = F,ylab = 'sd(r)',las=2, main='NB: Event-free')
boxplot(EI,outline = F,ylab = 'sd(r)',las=2, main='NB: Event-occuring')
dev.copy2pdf(file = 'ROutputFigs/GSE49711/boxplot_sd_preselected.pdf')

##################
## get network ###
load('outPut.Rdata/GSE49711/stable.top1relativeSD_morethan80.rData')

igraphL = getNetwork(A)
# LR-EFS:270 nodes
# ILR-EFS:209 nodes
# LR-E:62 nodes
# ILR-E:170 nodes
# HIR-EFS:114 nodes
# HR-EFS:152 nodes
# HIR-E:10 nodes
# HR-E:222 nodes

## note that using variable random walk steps may shift slightly the outputs
## the published resutls using step=1200 manually as an early work

cluster = getCluster(igraphL, step = 1200)
#cluster = getCluster_methods(igraphL) 

membersL_noweight = getMCI(cluster, A, adjust.size = F)  #getMCI(cluster.3,A, ylim = c(0,20))
#save(membersL_noweight,file = 'outPut.Rdata/GSE49711/membersL_noweight.nonormal.rData')  ?????

#par(mfrow = c(2,4))
plotBar_MCI(membersL_noweight,ylim = c(0,20),min = 2)
dev.copy2pdf(file = 'ROutputFigs/GSE49711/bar_MCI_top1relativeSD_nonormal_fdr0.05_noweighted.pdf')


## network plot #####
## original code: DNB_GSE49711_relativeMaxV2.R
#load("E:\\Zhezhen2019_bk\\GSE49711\\result\\HRLRm\\DNB\\GSE49711_only\\relativeMax\\test\\igraphL_maxCI_4stages.rData")

par(mfrow = c(1,2))
tmp = igraphL[["ILR-EFS"]]
E(tmp)$width <- E(tmp)$weight*3
V(tmp)$community= cluster[["ILR-EFS"]]$membership
mark.groups = table(cluster[["ILR-EFS"]]$membership)
colrs = rainbow(length(mark.groups), alpha = 0.3)
V(tmp)$label <- NA
plot(tmp, vertex.color=colrs[V(tmp)$community],vertex.size = 5,mark.groups=cluster[["ILR-EFS"]])

tmp = igraphL[[ "ILR-E"]]
E(tmp)$width <- E(tmp)$weight*3
V(tmp)$community= cluster[["ILR-E"]]$membership
mark.groups = table(cluster[["ILR-E"]]$membership)
colrs = rainbow(length(mark.groups), alpha = 0.3)
V(tmp)$label <- NA
plot(tmp, vertex.color=colrs[V(tmp)$community],vertex.size = 5,mark.groups=cluster[["ILR-E"]])


dev.copy2pdf(file ='ROutputFigs/GSE49711/network_critical_stages_outline.pdf')


######################################
## identify biomodule per state  #####
maxMCIms <- getMaxMCImember(membersL_noweight[["members"]], membersL_noweight[["MCI"]], min =30)

maxMCI = getMaxStats(membersL_noweight[['MCI']],maxMCIms[["idx"]])
maxMCI 
#LR-EFS   ILR-EFS     ILR-E   HIR-EFS    HR-EFS      HR-E 
#5.883866 19.105547 22.587388  7.815426  5.450432  6.882542 

## get biomodule with the highest MCI, CTS
CTS.E = getCTS(maxMCI, maxMCIms[["members"]])
#Length: 119
save(CTS.E, file = 'outPut.Rdata/GSE49711/Dataset1_GSE49711_E_CTS.rData')

x <- which(names(membersL_noweight[[1]]) %in% c('LR-E', 'ILR-E', 'HIR-E', 'HR-E'))
maxMCI.EF = getMaxStats(membersL_noweight[['MCI']][-x],maxMCIms[["idx"]][-x])
maxMCI.EF
#   LR-EFS   ILR-EFS   HIR-EFS    HR-EFS 
#  5.883866 19.105547  7.815426  5.450432 
CTS.EF = getCTS(maxMCI.EF, maxMCIms[["members"]][-x])
#Length: 137
save(CTS.EF, file = 'outPut.Rdata/GSE49711/Dataset2_GSE49711_EF_CTS.rData')

###########################################################
## check the significance of CTS by MCI scoring system  ###
## This step taks a while  ################################
## DO NOT RUN, because a severe evaluation is to test Ic ##
set.seed(2020)

x <- which(colnames(counts) %in% unlist(as.character(unlist(samplesL[c('LR-E', 'ILR-E', 'HIR-E', 'HR-E')]))))

simMCI.E = simulationMCI(length(CTS.E),samplesL[c('LR-E', 'ILR-E', 'HIR-E', 'HR-E')], counts[,x], B=2)

simMCI.EF = simulationMCI(length(CTS.EF),samplesL[c('LR-EF', 'ILR-EF', 'HIR-EF', 'HR-EF')], counts[,-x], B=2)

 

##################################################
### tipping point identification and evaluation ##
## original code: DNB_function_GSE49711_relativeMax.R
source("Rcode/GSE49711/GSE49711_E_BioTIP_plot_02242020.R")
source("Rcode/GSE49711/GSE49711_EF_BioTIP_plot_02242020.R")

## common motifs
load("outPut.Rdata/GSE49711/res+motif_GSE4only_E_hg19.rData")
tmp = as.data.frame(groupReport(res, top = 0.1))
tmp$fdr <- p.adjust(tmp$p.value,"fdr")
idx = row.names(tmp[tmp$raw.score>1 & tmp$top.motif.prop>0.15 & !grepl('UW.Motif.',tmp$target),])
tmp2 = tmp[idx,]
sig.E = subset(tmp2, fdr<0.05)
tmp3 = row.names(sig.E[!duplicated(sig.E$target),])
length(tmp3)  # 29
plot(groupReport(res,top = 0.1)[as.numeric(tmp3[1:length(tmp3)])],fontsize=5, id.fontsize=7)
dev.copy2pdf(file = 'ROutputFigs/GSE49711/motif_E_sig.pdf')

load("outPut.Rdata/GSE49711/res+motif_GSE4only_EF_hg19.rData")
tmp = as.data.frame(groupReport(res, top = 0.1))
tmp$fdr <- p.adjust(tmp$p.value,"fdr")
idx = row.names(tmp[tmp$raw.score>1 & tmp$top.motif.prop>0.15 & !grepl('UW.Motif.',tmp$target),])
tmp2 = tmp[idx,]
sig.EF = subset(tmp2, fdr<0.05)
tmp3 = row.names(sig.EF[!duplicated(sig.EF$target),])
length(tmp3)  # 23
plot(groupReport(res,top = 0.1)[as.numeric(tmp3[1:length(tmp3)])],fontsize=5, id.fontsize=7)
dev.copy2pdf(file = 'ROutputFigs/GSE49711/motif_EF_sig.pdf')


venn(list(E=unique(sig.E$target), EF=unique(sig.EF$target)))
intersect(unique(sig.E$target),unique(sig.EF$target))
#[1] "HNF4A"  "HNF4G"  "ATF6"   "DUS3L"  "EHF"    "NFIX"   "GPAM"   "PPP5C" 
#[9] "NR4A1"  "KIF22"  "NFE2L1" "MAX"    "HNF1B" 


