# Generated by: Xinan H Yang
# Revised by: Xinan H Yang (xyang2_at_uchicago.edu)
# Last update: 02/24/2020

setwd("F:/projects/BioTIP/doc/2020_/Applications/")

# The following initializes usage of Bioc devel
# BiocManager::install(version='devel')
# BiocManager::install("BioTIP")
library(BioTIP)
packageVersion("BioTIP")  #[1] 1.1.1

#source("../../../source/BioTIP_update_3.2_02242020.R")

set.seed(1138)
library(dplyr)
library(corpcor)
library("Biobase")

### DATA SET 1 ###
load("outPut.Rdata/GSE49711/Dataset1_GSE49711_E_CTS.rData")
length(CTS.E)  # 119
CTS <- CTS.E; rm(CTS.E)

load("input.Rdata/GSE49711/GSE49711_samplesL.rData")
samplesL <- samplesL[c('LR-E','ILR-E','HIR-E','HR-E')]
unlist(lapply(samplesL, length))
# HIR-E  HR-E ILR-E  LR-E 
#  11   120    29    23 

cli <- read.table('input.Rdata/GSE49711/full_cli_xy.txt' , header=T)
#tmp = names(table(cli$group_xy))

load("input.Rdata/GSE49711/full_table.simple.rData")
#load("E:\\ZheZhen2019_bk\\GSE49711\\result\\HRLRm\\full_table.MergedTable_methy.rData")
#full_table <- full_table[,1:17]w
expressed = subset(full_table, AveExpr>1)
dim(expressed)  # [1] 48682   17

load("input.Rdata/GSE49711/eSet_log2.rData")
count_matrix = exprs(eSet)
rm(eSet)
count_matrix = subset(count_matrix,rownames(count_matrix) %in% expressed$Row.names)
dim(count_matrix) #[1] 48682   498

count_matrix = count_matrix[,which(colnames(count_matrix) %in% as.character(unlist(samplesL)))]
dim(count_matrix) #[1] 48682   183


##################################################
### Andrew's scratch ###########  
# I_c = function(cor_G, cor_S) {
#   num = mean(abs(cor_G[lower.tri(cor_G)]), na.rm = TRUE)
#   denom = mean(cor_S[lower.tri(cor_S)], na.rm = TRUE)
#   return(num / denom)
# }
# 
# # count_matrix is G x S (row for each gene, column for each sample)
# compute_I_c = function(count_matrix, shrink = TRUE) {
#   if (shrink) {
#     cor_genes = cor.shrink(count_matrix, MARGIN=1, target='zero')
#     cor_samples = cor.shrink(count_matrix, MARGIN=2, target='average')
#   } else {
#     cor_genes = cor(t(count_matrix))
#     cor_samples = cor(count_matrix)
#   }
#   return(I_c(cor_genes, cor_samples))
# }
# 
# type = unlist(samplesL)
# type = as.character(type)
# type = cbind(type, rep(names(samplesL), times = sapply(samplesL, length)))
# 
# count_matrix_relevant = count_matrix[CTS, ]
# 
# count_matrix_relevant_LR_E = count_matrix_relevant[, as.character(samplesL$`LR-E`)]
# 
# # LR-E
# (I_c_LR = compute_I_c(count_matrix_relevant_LR_E, shrink = F))  # 0.2288027
# (I_c_LR_shrink = compute_I_c(count_matrix_relevant_LR_E, shrink = T)) # 110.3237
# 


##################################################
######  BioTIP score, shulffing genes ##############
  C= 1000
  n <- length(CTS)
  BioTIP_score <- getIc(count_matrix, samplesL, CTS, fun="BioTIP", shrink=TRUE, PCC_sample.target='average')
  BioTIP_score
  #    LR-E     ILR-E     HIR-E      HR-E 
  # 0.04056869 0.86224462 0.04276787 0.14576600!
  
  simuBioTIP_g  <- simulation_Ic(n, samplesL, count_matrix, B=C,
                                  fun="BioTIP", shrink=TRUE)
  
#### BioTIP score,  shulffing labelling
    x <- 2
  #  ns <- length(samplesL[[x]])  # of cells at the state of interest
    simuBioTIP_s <- matrix(nrow=length(samplesL), ncol=C)
    rownames(simuBioTIP_s) = names(samplesL)
    for(j in 1:length(samplesL)) {
      ns <- length(samplesL[[j]])  # of each state
      simuBioTIP_s[j,] <- simulation_Ic_sample(count_matrix, ns, Ic=BioTIP_score[x],
                                                 genes=CTS, B=C,
                                                 fun="BioTIP", shrink=TRUE)
    }
 # save( simuBioTIP_g,  simuBioTIP_s, file="F:/projects/BioTIP/result/GSE49711/E_simuBioTIP.RData", compress=TRUE)
    
 # load(file="F:/projects/BioTIP/result/GSE49711/E_simuBioTIP.RData")  
  pdf(file="ROutputFigs/GSE49711/Dataset1_E.matplot_density_Ic_simulation.pdf",width=10, height=7)  
  par(mfrow=c(1,2))
      ## global matplot 
  plot_Ic_Simulation(BioTIP_score, simuBioTIP_g, las = 2, ylab="BioTIP",
                       main="CTS.E (119 transcripts)",
                       fun="matplot", which2point= 'ILR-E')
    ## #local Kernel Density Plot  
  d <- density(simuBioTIP_s['ILR-E',]) # returns the density data
  plot(d) # plots the results
  abline(v=BioTIP_score['ILR-E'], col="green")
  dev.off() 
   
     
   ## supporting GitHub plots ###      
   par(mfrow=c(2,2))
  
  plot_Ic_Simulation(BioTIP_score, simuBioTIP_g, las = 2, ylab="BioTIP",
                     main=paste(length(CTS),names(samplesL)[x], "genes", "\n","vs. ",C, "gene-permutations"),
                     fun="boxplot", which2point= x)
  plot_SS_Simulation(BioTIP_score, simuBioTIP_g, 
                     main = paste("Delta BioTIP",n,"genes"), 
                     ylab=NULL)
  plot_Ic_Simulation(BioTIP_score, simuBioTIP_s, las = 2, ylab="BioTIP", ylim=c(-0.05, 1),
                     main=paste(length(CTS), names(samplesL)[x],"genes", "\n","vs. ",C, "sample-permutations"),
                     fun="matplot", which2point= x)
  plot_SS_Simulation(BioTIP_score, simuBioTIP_s, 
                     main = "Delta BioTIP, label shuffling", 
                     ylab=NULL)
  
  dev.copy2pdf(file=paste0("ROutputFigs/GSE49711/Dataset1_E_BioTIP_vsSimulation_",length(CTS),"CTS.pdf"))



  ##################################################
  ######  Ic score, of the same number of random genes ##############
  C= 1000
  n <- length(CTS)
  Ic_score <- getIc(count_matrix, samplesL, CTS, fun="cor", PCC_sample.target='average')
  Ic_score
 # LR-E     ILR-E     HIR-E      HR-E 
 # 0.2352912 1.0954868 0.3256219 0.2478748 
  simuIc_g  <- simulation_Ic(n, samplesL, count_matrix, B=C,
                                 fun='cor')
  
  
  #### Ic score for the identified CTS,  shulffing labelling
  x <- 2
#  ns <- length(samplesL[[x]])  # of fixed number of cells
  simuIc_s <- matrix(nrow=length(samplesL), ncol=C)
  rownames(simuIc_s) = names(samplesL)
  for(j in 1:length(samplesL)) { 
    ns <- length(samplesL[[j]])  # for each state rewpectively 
    simuIc_s[j,] <- simulation_Ic_sample(count_matrix, ns, Ic=Ic_score[x],
                                             genes=CTS, B=C,
                                             fun="cor")
  }
  # save( simuIc_g,  simuIc_s, file="F:/projects/BioTIP/result/GSE49711/E_simuIC.RData", compress=TRUE)
  
  # load(file="F:/projects/BioTIP/result/GSE49711/E_simuIC.RData")  
  
  
  par(mfrow=c(2,2))
  
  plot_Ic_Simulation(Ic_score, simuIc_g, las = 2, ylab="Ic",
                     main=paste(length(CTS),names(samplesL)[x], "genes", "\n","vs. ",C, "gene-permutations"),
                     fun="boxplot", which2point= x)
  plot_SS_Simulation(Ic_score, simuIc_g, 
                     main = paste("Delta Ic",n,"genes"), 
                     ylab=NULL)
  plot_Ic_Simulation(Ic_score, simuIc_s, las = 2, ylab="Ic", ylim=c(-0.01, 1.2),
                     main=paste(length(CTS), names(samplesL)[x],"genes", "\n","vs. ",C, "sample-permutations"),
                     fun="matplot", which2point= x)
  plot_SS_Simulation(Ic_score, simuIc_s, 
                     main = "Delta BioTIP, label shuffling", 
                     ylab=NULL)
  
  dev.copy2pdf(file=paste0("ROutputFigs/GSE49711/Dataset1_E_Ic_vsSimulation_",length(CTS),"CTS.pdf"))
  


  
  ########################
  ## motif analysis
library("Biobase")
library(gplots)
library(GenomicRanges)
library(BSgenome.Hsapiens.UCSC.hg19)
library(PWMEnrich.Hsapiens.background)
data(PWMLogn.hg19.MotifDb.Hsap)
motifs <- PWMLogn.hg19.MotifDb.Hsap


source('Rcode/functions_motif_analysis.R')

load("outPut.Rdata/GSE49711/Dataset1_GSE49711_E_CTS.rData")
length(CTS.E) #[1] 119  

load("input.Rdata/GSE49711/full_table.simple.rData")
dim(full_table)   # [1] 60778    17

CTS_coord = GRanges(subset(full_table, Row.names %in% CTS.E)$ID)
CTS_gr = promoters(GRanges(CTS_coord),upstre = 1500,downstream = 500)
 
blacklist = read.delim("input.Rdata/GSE49711/hg19-blacklist.v2.bed",
sep ='\t',head = F)
colnames(blacklist) = c('chr','start','end','name')
blacklist = GRanges(blacklist)

CTS_gr = setdiff(CTS_gr, subsetByOverlaps(CTS_gr,blacklist))
length(CTS_gr) #[1] 119

res <- motif_analysis(CTS_gr, BSgenome.Hsapiens.UCSC.hg19,
       saveN='GSE4only_E_hg19',motifs,save = TRUE, path='outPut.Rdata/GSE49711/')
 


  
  
  
 