# Generated by: Xinan H Yang
# Revised by: Xinan H Yang (xyang2_at_uchicago.edu)
# Last update: 02/24/2020

setwd("F:/projects/BioTIP/doc/2020_/Applications/")

# The following initializes usage of Bioc devel
BiocManager::install(version='devel')
BiocManager::install("BioTIP")
library(BioTIP)
packageVersion("BioTIP")  #[1] 1.1.1

#source("../../../source/BioTIP_update_3.2_02242020.R")

set.seed(1138)
library(dplyr)
library(corpcor)
library(Biobase)



#######################################
#######################################
### DATA SET 2 ###
load("outPut.Rdata/GSE49711/Dataset2_GSE49711_EF_CTS.rData")
length(CTS.EF)  # 137
CTS <- CTS.EF; rm(CTS.EF)


load("input.Rdata/GSE49711/GSE49711_samplesL.rData")
samplesL <- samplesL[c('LR-EFS','ILR-EFS','HIR-EFS','HR-EFS')]
unlist(lapply(samplesL, length))
#LR-EFS ILR-EFS HIR-EFS  HR-EFS 
# 168      72      19      56 

cli <- read.table('input.Rdata/GSE49711/full_cli_xy.txt' , header=T)
#tmp = names(table(cli$group_xy))

load("input.Rdata/GSE49711/full_table.simple.rData")
#load("E:\\ZheZhen2019_bk\\GSE49711\\result\\HRLRm\\full_table.MergedTable_methy.rData")
#full_table <- full_table[,1:17]w
expressed = subset(full_table, AveExpr>1)
dim(expressed)  # [1] 48682   17

load("input.Rdata/GSE49711/eSet_log2.rData")
count_matrix = exprs(eSet)
rm(eSet)
count_matrix = subset(count_matrix,rownames(count_matrix) %in% expressed$Row.names)
dim(count_matrix) #[1] 48682   498

count_matrix = count_matrix[,which(colnames(count_matrix) %in% as.character(unlist(samplesL)))]
dim(count_matrix) #[1] 48682   315


##################################################
######  BioTIP score, shulffing genes ##############
C= 1000
(n <- length(CTS)) # 137
BioTIP_score <- getIc(count_matrix, samplesL, CTS, fun="BioTIP", PCC_sample.target='average')
BioTIP_score
# LR-EFS    ILR-EFS    HIR-EFS     HR-EFS 
#0.04698274 0.66304154 0.09631273 0.05663618 

simuBioTIP_g  <- simulation_Ic(n, samplesL, count_matrix, B=C,
                               fun="BioTIP")


#### BioTIP score,  shulffing labelling
x <- 2
#ns <- length(samplesL[[x]])  # of cells
simuBioTIP_s <- matrix(nrow=length(samplesL), ncol=C)
rownames(simuBioTIP_s) = names(samplesL)
for(j in 1:length(samplesL)) {
  ns <- length(samplesL[[j]])
  simuBioTIP_s[j,] <- simulation_Ic_sample(count_matrix, ns, Ic=BioTIP_score[x],
                                           genes=CTS, B=C,
                                           fun="BioTIP")
}
#  save( simuBioTIP_g,  simuBioTIP_s, file="F:/projects/BioTIP/result/GSE49711/EF_simuBioTIP.RData", compress=TRUE)

# load(file="F:/projects/BioTIP/result/GSE49711/EF_simuBioTIP.RData")  
 
 pdf(file="ROutputFigs/GSE49711/Dataset2_EF.matplot_density_Ic_simulation.pdf",width=10,height=7) 
 par(mfrow=c(1,2))
      ## global matplot 
  plot_Ic_Simulation(BioTIP_score, simuBioTIP_g, las = 2, ylab="BioTIP",
                       main="CTS.EF (137 transcripts)", 
                       fun="matplot", which2point= "ILR-EFS")
    ## #local Kernel Density Plot  
  d <- density(simuBioTIP_s['ILR-EFS',]) # returns the density data
  plot(d) # plots the results
  abline(v=BioTIP_score['ILR-EFS'], col="green")
  dev.off() 

par(mfrow=c(2,2))

plot_Ic_Simulation(BioTIP_score, simuBioTIP_g, las = 2, ylab="BioTIP",  ylim=c(0.05,0.2),
                   main=paste(length(CTS),names(samplesL)[x], "genes", "\n","vs. ",C, "gene-permutations"),
                   fun="boxplot", which2point= x)
plot_SS_Simulation(BioTIP_score, simuBioTIP_g, 
                   main = paste("Delta BioTIP",n,"genes"), 
                   ylab=NULL)
plot_Ic_Simulation(BioTIP_score, simuBioTIP_s, las = 2, ylab="BioTIP", ylim=c(0.01, 0.8),
                   main=paste(length(CTS), names(samplesL)[x],"genes", "\n","vs. ",C, "sample-permutations"),
                   fun="matplot", which2point= x)
plot_SS_Simulation(BioTIP_score, simuBioTIP_s, 
                   main = "Delta BioTIP, label shuffling", 
                   ylab=NULL)

dev.copy2pdf(file=paste0("ROutputFigs/GSE49711/Dataset2_EF_BioTIP_vsSimulation_",length(CTS),"CTS.pdf"))

wilcox.test(simuBioTIP_g[1,],simuBioTIP_g[2,]) #P,2e-16
wilcox.test(simuBioTIP_g[3,],simuBioTIP_g[2,]) #P,2e-16


##################################################
######  Ic score, shulffing genes ##############
C= 1000
n <- length(CTS)
Ic_score <- getIc(count_matrix, samplesL, CTS, fun="cor", PCC_sample.target='average')
Ic_score
#LR-EFS   ILR-EFS   HIR-EFS    HR-EFS 
#0.1081278 0.7610321 0.7172504 0.1799052 
simuIc_g  <- simulation_Ic(n, samplesL, count_matrix, B=C,
                           fun='cor')


#### Ic score,  shulffing labelling
x <- 2
#ns <- length(samplesL[[x]])  # of cells
simuIc_s <- matrix(nrow=length(samplesL), ncol=C)
rownames(simuIc_s) = names(samplesL)
for(j in 1:length(samplesL)) {
  ns <- length(samplesL[[j]]) 
  simuIc_s[j,] <- simulation_Ic_sample(count_matrix, ns, Ic=Ic_score[x],
                                       genes=CTS, B=C,
                                       fun="cor")
}

# save( simuIc_g,  simuIc_s, file="F:/projects/BioTIP/result/GSE49711/EF_simuIC.RData", compress=TRUE)

# load(file="F:/projects/BioTIP/result/GSE49711/EF_simuIC.RData")  

par(mfrow=c(2,2))

plot_Ic_Simulation(Ic_score, simuIc_g, las = 2, ylab="Ic",
                   main=paste(length(CTS),names(samplesL)[x], "genes", "\n","vs. ",C, "gene-permutations"),
                   fun="boxplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_g, 
                   main = paste("Delta Ic",n,"genes"), 
                   ylab=NULL)
plot_Ic_Simulation(Ic_score, simuIc_s, las = 2, ylab="Ic",ylim=c(0.01,0.8),
                   main=paste(length(CTS), names(samplesL)[x],"genes", "\n","vs. ",C, "sample-permutations"),
                   fun="matplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_s, 
                   main = "Delta Ic, label shuffling",  
                   ylab=NULL)

dev.copy2pdf(file=paste0("ROutputFigs/GSE49711/Dataset2_EF_Ic_vsSimulation_",length(CTS),"CTS.pdf"))



######################################################################
######  Ic score, shulffing genes, 2nd run, the SAME !!!##############
# C= 1000
# n <- length(CTS)
# Ic_score <- getIc(count_matrix, samplesL, CTS, fun="BioTIP", PCC_sample.target='none', PCC_gene.target='none')
# Ic_score
# # LR-EF     ILR-EF     HIR-EF      HR-EF 
# # 0.1081278 0.7610321 0.7172504 0.1799052 
# simuIc_g  <- simulation_Ic(n, samplesL, count_matrix, B=C,
#                            fun="BioTIP")
# 
# 
# #### Ic score,  shulffing labelling
# x <- 2
# ns <- length(samplesL[[x]])  # of cells
# simuIc_s <- matrix(nrow=length(samplesL), ncol=C)
# rownames(simuIc_s) = names(samplesL)
# for(j in 1:length(samplesL)) {
#   simuIc_s[j,] <- simulation_Ic_sample(count_matrix, ns, Ic=Ic_score[x],
#                                        genes=CTS, B=C,
#                                        fun="BioTIP")
# }
# 
# 
# par(mfrow=c(2,2))
# 
# plot_Ic_Simulation(Ic_score, simuIc_g, las = 2, ylab="Ic",
#                    main=paste(length(CTS),names(samplesL)[x], "genes", "\n","vs. ",C, "gene-permutations"),
#                    fun="boxplot", which2point= x)
# plot_SS_Simulation(Ic_score, simuIc_g, 
#                    main = paste("Larget - 2nd largest",ns,"samples"), 
#                    ylab="Ic")
# plot_Ic_Simulation(Ic_score, simuIc_s, las = 2, ylab="Ic",
#                    main=paste(length(CTS), names(samplesL)[x],"genes", "\n","vs. ",C, "sample-permutations"),
#                    fun="matplot", which2point= x)
# plot_SS_Simulation(Ic_score, simuIc_s, 
#                    main = paste("Larget - 2nd largest",ns,"samples"), 
#                    ylab="Ic")
# 
# dev.copy2pdf(file=paste0("ROutputFigs/GSE49711/Dataset2_EF_Ic_vsSimulation_",length(CTS),"CTS_2nd.pdf"))

  ########################
  ## motif analysis
library("Biobase")
library(gplots)
library(GenomicRanges)
library(BSgenome.Hsapiens.UCSC.hg19)
library(PWMEnrich.Hsapiens.background)
data(PWMLogn.hg19.MotifDb.Hsap)
motifs <- PWMLogn.hg19.MotifDb.Hsap


source('Rcode/functions_motif_analysis.R')

load("outPut.Rdata/GSE49711/Dataset2_GSE49711_EF_CTS.rData")
length(CTS.EF) #[1] 137  

load("input.Rdata/GSE49711/full_table.simple.rData")
dim(full_table)   # [1] 60778    17

CTS_coord = GRanges(subset(full_table, Row.names %in% CTS.EF)$ID)
CTS_gr = promoters(GRanges(CTS_coord),upstre = 1500, downstream = 500)
 
blacklist = read.delim("input.Rdata/GSE49711/hg19-blacklist.v2.bed",
sep ='\t',head = F)
colnames(blacklist) = c('chr','start','end','name')
blacklist = GRanges(blacklist)

CTS_gr = setdiff(CTS_gr, subsetByOverlaps(CTS_gr,blacklist))
length(CTS_gr) #[1] 132

res <- motif_analysis(CTS_gr, BSgenome.Hsapiens.UCSC.hg19,
       saveN='GSE4only_EF_hg19',motifs,save = TRUE, path='outPut.Rdata/GSE49711/')
 

