# This code performs CTS analysis on the log2-transformed gene expressional data of GSE6136, using BioTIP (v.1.1.1).
# Generated by: Zhezhen Wang
# Revised by: Xinan H Yang (xyang2_at_uchicago.edu)
# Last update: 04/03/2020

# The following initializes usage of Bioc devel
BiocManager::install(version='devel')
BiocManager::install("BioTIP")
library(BioTIP)
packageVersion("BioTIP")  #[1] '1.1.1'


library(stringr)
library(psych)
library(igraph)


# instead of the following one command
#source("F:/projects/BioTIP/source/BioTIP_update_3.3_02282020.R")  # combined functions with parameters;
 setwd("F:/projects/BioTIP/doc/2020_/Applications/")

####################### load in the downloaded expressinal matrix of GSE6136 #####################
 GSE6136 = read.table("input.Rdata/GSE6136/GSE6136_matrix.txt",head = T)
 dim(GSE6136) #[1] 22690    27
 row.names(GSE6136) = GSE6136$ID_REF
 GSE6136 = as.matrix(GSE6136[,-1])
 dim(GSE6136) #[1] 22690    26
 range(GSE6136)     # check if log-transformed
# [1]     0.0 80843.2

 ## read the published sample states
 cli = read.delim("input.Rdata/GSE6136/GSE6136_cli.txt",head = F)
 cli = t(cli)
 library(stringr)
 colnames(cli) = str_split_fixed(cli[1,],'_',2)[,2]
 cli = cli[-1,]
 cli = data.frame(cli)
 cli[,'cell-type:ch1'] = str_split_fixed(cli$characteristics_ch1.1,': ',2)[,2]
 cli[,'Ig clonality:ch1'] = str_split_fixed(cli$characteristics_ch1.3,': ',2)[,2]

 table(cli[,'cell-type:ch1'])
#            activated (P2)  lymphoma_aggressive (P5)    lymphoma_marginal (P3)
#                    3                     7                     6
# lymphoma_transitional (P4)              resting (P1)
#                   5                     5
table(cli[,'characteristics_ch1'])
#genotype: E-mu-BRD2  genotype: wildtype 
#                 21                   5 
table(cli[,'Ig clonality:ch1'])
# monoclonal oligoclonal  polyclonal
#          7          11           8

####################### BioTIP application  #####################

cli$group = cli$'cell-type:ch1'
df = log2(GSE6136+1)
tmp = names(table(cli$group))
samplesL = split(cli$geo_accession, f = cli$group)

## step 1) pre-select the transcrips
## optimation is ignored in this trial due to small sample size
table(cli$group)
#activated   lymphoma_aggressive     lymphoma_marginal lymphoma_transitional               resting
#    3                     7                     6                     5                     5

## Preselect the top 5% transcripts with the highest relative standard deviation
## in each state compared to transcript in the other states
subcounts = sd_selection(df,samplesL,cutoff= 0.05)
sapply(subcounts,dim)
#     activated lymphoma_aggressive lymphoma_marginal lymphoma_transitional resting
#[1,]      1134                1134              1134                  1134    1134
#[2,]         3                   7                 6                     5       5

#pre5s_5per = row.names(subcounts[['activated']])
#save(pre5s_5per, file = 'F:/projects/BioTIP/result/GSE6136_Zhezhen/Holly_desktop/pre5s_5per_preselect_GSM142418.rData')

## step 2.1) Build gene regulatory network based on expresional PCC
igraphL = getNetwork(subcounts,fdr=1)
#activated:1134 nodes
#lymphoma_aggressive:1134 nodes
#lymphoma_marginal:1134 nodes
#lymphoma_transitional:1134 nodes
#resting:1134 nodes

## step 2.2) Network partition using random walk
## this step constructs modules (geoups of transcriopts) based on their expression levels
cluster = getCluster_methods(igraphL)

## step 3.1) calculate MCI score for each module
membersL_noweight = getMCI(cluster, subcounts, adjust.size = F)
par(mfrow = c(1,4))
plotBar_MCI(membersL_noweight,ylim = c(0,4),min = 30)
dev.copy2pdf(file = 'ROutputFigs/GSE6136/MCI_GSE6136_5states_5per.pdf')

## step 3.2) identify biomodule per state
maxMCIms <- getMaxMCImember(membersL_noweight[[1]],membersL_noweight[[2]],min =30)

maxMCI = getMaxStats(membersL_noweight[['MCI']],maxMCIms[["idx"]])

## get biomodule with the highest MCI, CTS
CTS = getCTS(maxMCI, maxMCIms[["members"]])
#Length: 141
#save(CTS, file = 'outPut.Rdata/GSE6136_robust_analysis/CTS.GSE6136_5states_5per.rData')

## step 3.3) check the significance of CTS by MCI scoring system 
set.seed(2020)
simMCI = simulationMCI(length(CTS),samplesL, df)

## step 4.1) Caslculate the ic-score and evaluate its significance
## permulation genes (gene-size controled)
Ic = getIc(df,samplesL,CTS)
simIc = simulation_Ic(length(CTS),samplesL, df)

save(maxMCI, simMCI, simIc, Ic, file='outPut.Rdata/GSE6136/MCI_Ic_genepermutation_GSE6136_BioTIP141g.RData', compress=TRUE)

## plot 
par(mfrow = c(1,2))
plot_MCI_Simulation(maxMCI,simMCI,las = 2,
  order = c('resting','activated','lymphoma_marginal','lymphoma_transitional','lymphoma_aggressive'),
  main= paste("BioTIP", length(CTS), "1000 gene permutations"))

plot_Ic_Simulation(Ic,simIc,
   order = c('resting','activated','lymphoma_marginal','lymphoma_transitional','lymphoma_aggressive'),
   main= paste("BioTIP", length(CTS), "1000 gene permutations"))

dev.copy2pdf(file = 'ROutputFigs/GSE6136/MCI_Ic_GSE6136_5states_5per.pdf')

## step 4.2) further check the significance of CTS by new Ic scores
## permulation genes (gene-size controled)
Ic_adjusted = getIc(df,samplesL,CTS, fun ='BioTIP')
simIc_adjusted = simulation_Ic(length(CTS),samplesL, df,  fun ='BioTIP')
plot_Ic_Simulation(Ic_adjusted, simIc_adjusted, ylim = c(0,6), las=2, ylab='adjusted Ic',
   order = c('resting','activated','lymphoma_marginal','lymphoma_transitional','lymphoma_aggressive'),
   main= paste("GSE6136", length(CTS), "1000 gene permutations") )
#dev.copy2pdf(file = 'ROutputFigs/GSE6136_robust_analysis/MCI_Ic_GSE6136_5states_5per.pdf')





####################################################
## Compare to the 27 genes reported by Chen 2012  ##
####################################################
dnb_chen = read.table("input.Rdata/GSE6136/DNB_Chen2011_xy.txt",head = T)
dnb_chen <- as.vector(dnb_chen$ID)
all(as.vector(dnb_chen) %in% rownames(df))   #[1] TRUE

set.seed(2020)
simMCI = simulationMCI(length(dnb_chen),samplesL, df, fun="BioTIP", shrink=FALSE)

Ic = getIc(df,samplesL,dnb_chen)
simIc = simulation_Ic(length(dnb_chen),samplesL, df)
save(simMCI, simIc, Ic, file='outPut.Rdata/GSE6136/MCI_Ic_genepermutation_log2GSE6136_Chen27g.RData', compress=TRUE)

x = "activated"
intersect(dnb_chen, rownames(subcounts[[x]])) # 0 chen's SNB genes has been preselected by sd() in the 'activated' state

m = df[dnb_chen, samplesL[[x]]]
comple = df[-which(rownames(df) %in% dnb_chen),samplesL[[x]]]

PCCo = abs(cor(t(comple),t(m)))
PCCo_avg = mean(PCCo, na.rm=TRUE) 
#PCCo_avg = avg.cor.shrink(comple, m, MARGIN = 1,  shrink = FALSE)
PCCi = abs(cor(t(m)))
PCCi_avg = (sum(PCCi, na.rm = T) - nrow(PCCi))/(nrow(PCCi)^2-nrow(PCCi)) 
#PCCi_avg = avg.cor.shrink(m, MARGIN = 1,  shrink = FALSE)
sd_chen = apply(m, 1, sd)
MCI_chen = mean(sd_chen)*(PCCi_avg/PCCo_avg)
names(MCI_chen) <- x

par(mfrow = c(1,2))
plot_MCI_Simulation(MCI_chen, simMCI, las = 2, 
   order = c('resting','activated','lymphoma_marginal','lymphoma_transitional','lymphoma_aggressive'),
   main= paste("Chen 2017", length(dnb_chen), "1000 gene permutations") )

plot_Ic_Simulation(Ic,simIc, las=2, ylab='adjusted Ic',fun="boxplot",
   order = c('resting','activated','lymphoma_marginal','lymphoma_transitional','lymphoma_aggressive'),
   main= paste("Chen 2017", length(dnb_chen), "1000 gene permutations") )

dev.copy2pdf(file = 'ROutputFigs/GSE6136/MCI_Ic_log2GSE6136_Chen27g.pdf')

##### Alternatively, test on the non-log-transformed data ###################
all(as.vector(dnb_chen) %in% rownames(GSE6136))   #[1] TRUE

simMCI = simulationMCI(length(dnb_chen),samplesL, GSE6136)

Ic = getIc(GSE6136,samplesL,dnb_chen)
simIc = simulation_Ic(length(dnb_chen),samplesL, GSE6136)
save(simMCI, simIc, Ic, file='outPut.Rdata/GSE6136/MCI_Ic_genepermutation_GSE6136_Chen27g.RData', compress=TRUE)

x = "activated"  
dim(subcounts[[x]])  # 1134  3
intersect(rownames(subcounts[[x]] ), dnb_chen)  # 0

m = GSE6136[unique(dnb_chen), samplesL[[x]]]
comple = GSE6136[setdiff(rownames(GSE6136), dnb_chen), samplesL[[x]]]

PCCo = abs(cor(t(comple),t(m)))
PCCo_avg = mean(PCCo, na.rm=TRUE) 
PCCi = abs(cor(t(m)))
PCCi_avg = (sum(PCCi, na.rm = T) - nrow(PCCi))/(nrow(PCCi)^2-nrow(PCCi)) 
sd_chen = apply(m, 1, sd)
MCI_chen = mean(sd_chen)*(PCCi_avg/PCCo_avg)
names(MCI_chen) <- x  # 154

par(mfrow = c(1,2))
plot_MCI_Simulation(MCI_chen, simMCI, las = 2,
   order = c('resting','activated','lymphoma_marginal','lymphoma_transitional','lymphoma_aggressive'),
   main= paste("Chen 2017", length(dnb_chen), "1000 gene permutations") )

plot_Ic_Simulation(Ic,simIc, las=2, ylab='adjusted Ic',fun="boxplot",
   order = c('resting','activated','lymphoma_marginal','lymphoma_transitional','lymphoma_aggressive'),
   main= paste("Chen 2017", length(dnb_chen), "1000 gene permutations") )

dev.copy2pdf(file = 'ROutputFigs/GSE6136/MCI_Ic_GSE6136_Chen27g.pdf')

wilcox.test(simIc[1,], simIc[2,])  P<2e-16
wilcox.test(simIc[3,], simIc[2,])  P<2e-16
wilcox.test(simIc[4,], simIc[2,])  P<2e-16




