# This code performs CTS analysis on the log2-transformed gene expressional data of GSE6136, using BioTIP (v.1.1.1).
# Generated by: Zhezhen Wang
# Revised by: Xinan H Yang (xyang2_at_uchicago.edu), Yuxi Sun (ysun11_at_uchicago.edu)
# Last update: 09/30/2020

# The following initializes usage of Bioc devel
# BiocManager::install(version='devel')
# BiocManager::install("BioTIP")
#source("/Users/jennifersun/Desktop/BioTIP/R/BioTIP-package.R")
#source("/Users/jennifersun/Desktop/BioTIP/R/BioTIP_update_3.3_02282020.R")
#source("/Users/jennifersun/Desktop/BioTIP/R/data.R")
#source("F:/projects/BioTIP/R_function/BioTIP-master/R/BioTIP_update_4_09282020.R")

library(BioTIP)
packageVersion("BioTIP")  #[1] 'v. 1.3.0'

library(stringr)
library(psych)
library(igraph)

####################### load in the downloaded expressinal matrix of GSE6136 #####################
#setwd("/Users/jennifersun/Desktop/BioTIP_Applications/input.Rdata/GSE6136")
setwd("F:/projects/BioTIP/doc/2020_/Applications/")
GSE6136 = read.table("input.Rdata/GSE6136/GSE6136_matrix.txt", head = T)
dim(GSE6136) #[1] 22690    27
row.names(GSE6136) = GSE6136$ID_REF
GSE6136 = GSE6136[,-1]
dim(GSE6136) #[1] 22690    26
GSE6136 <- as.matrix(GSE6136)

## read the published sample states
cli = read.delim("input.Rdata/GSE6136/GSE6136_cli.txt",head = F)
cli = t(cli)
colnames(cli) = str_split_fixed(cli[1,],'_',2)[,2]
cli = cli[-1,]
cli = data.frame(cli)
cli[,'cell-type:ch1'] = str_split_fixed(cli$characteristics_ch1.1,': ',2)[,2]
cli[,'Ig clonality:ch1'] = str_split_fixed(cli$characteristics_ch1.3,': ',2)[,2]

table(cli[,'cell-type:ch1'])
#            activated (P2)  lymphoma_aggressive (P5)    lymphoma_marginal (P3)
#                    3                     7                     6
#lymphoma_transitional (P4)              resting (P1)
#                   5                     5
table(cli[,'characteristics_ch1'])
#genotype: E-mu-BRD2  genotype: wildtype 
#                 21                   5 
table(cli[,'Ig clonality:ch1'])
# monoclonal oligoclonal  polyclonal
#          7          11           8

####################### BioTIP application  #####################
## as long as the paackage BioTIP is availabel from Bioconductor, you will use
# library(BioTIP)
# instead of the following one command

cli$group = cli$'cell-type:ch1'

df = log2(GSE6136+1)
tmp = names(table(cli$group))
cli$group = factor(cli$group, levels=c('resting','activated','lymphoma_marginal','lymphoma_transitional','lymphoma_aggressive'))
samplesL = split(cli$geo_accession,f = cli$group)

s## step 2.1) pre-select the transcrips
## optimation is ignored in this trial due to small sample size
table(cli$group)
#            activated   lymphoma_aggressive     lymphoma_marginal lymphoma_transitional               resting
#                    3                     7                     6                     5                     5

## Instead, preselect the top 1% transcripts with the highest relative standard deviation
## in each state compared to transcript in the other states
subcounts = sd_selection(df,samplesL,cutoff= 0.01)
sapply(subcounts,dim)
#     activated lymphoma_aggressive lymphoma_marginal lymphoma_transitional resting
#[1,]      1134                1134              1134                  1134    1134
#[2,]         3                   7                 6                     5       5

pre5s_1per = row.names(subcounts[['activated']])
#save(pre5s_5per, file = 'pre5s_1per_preselect_GSM142418.rData')

## step 3.1) Build gene regulatory network based on expresional PCC
igraphL = getNetwork(subcounts,fdr=1)

## step 3.2) Network partition using random walk
## this step constructs modules (geoups of transcriopts) based on their expression levels
cluster = getCluster_methods(igraphL)

## step 4.1-4.3) calculate MCI score for each module
membersL_noweight = getMCI(cluster,subcounts,adjust.size = F)
par(mar=c(1,1,1,1))
plotBar_MCI(membersL_noweight,ylim = c(0,4),min = 30)


dev.copy2pdf(file = 'ROutputFigs/GSE6136/bar_GSE6136_5states_1per.pdf')

## step 4.4) identify biomodule per state
maxMCIms <- getMaxMCImember(membersL_noweight[[1]],membersL_noweight[[2]],min =30)

maxMCI = getMaxStats(membersL_noweight[['MCI']],maxMCIms[[1]])

## get biomodule with the highest MCI, CTS
CTS = getCTS(maxMCI, maxMCIms[[2]])

#Length: 141(5), 35(1)
save(CTS, file = 'outPut.Rdata/GSE6136_robust_analysis/CTS.GSE6136_5states_1per.rData')


###########################################################
#par(mfrow = c(2,1))
#simMCI = simulationMCI(105,samplesL, df)
#plot_MCI_Simulation(maxMCI,simMCI,ylim =c(0,5),las = 2,order = c('resting','activated','lymphoma_marginal','lymphoma_transitional','lymphoma_aggressive'))
load(file = 'outPut.Rdata.bk/GSE6136_robust_analysis/CTS.GSE6136_5states_1per.rData')
#

Ic = getIc(df,samplesL,CTS,fun="BioTIP")
simIc = simulation_Ic(length(CTS),samplesL, df, fun="BioTIP")

plot_Ic_Simulation(Ic,simIc,
                   order = c('resting','activated','lymphoma_marginal','lymphoma_transitional','lymphoma_aggressive'),
                   las = 2, ylab = "Ic*",
                   main= paste("BioTIP", length(CTS), "1000 gene permutations"), fun='boxplot')
save(maxMCI, simMCI, simIc, Ic, file='outPut.Rdata/GSE6136/MCI_Ic_genepermutation_GSE6136_BioTIP141g_09292020.RData', compress=TRUE)


dev.copy2pdf(file = 'ROutputFigs/GSE6136/MCI_Ic_GSE6136_5states_1per.pdf')
